{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate binned methylation around MEVs overlapping genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyarrow\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "import os,re\n",
    "from multiprocessing import Pool\n",
    "from tqdm import tqdm\n",
    "from itertools import repeat\n",
    "\n",
    "import sys\n",
    "# add path to src module - note this is not a Pythonic solution\n",
    "sys.path.insert(1, '../') \n",
    "\n",
    "from src import calculate_mC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "me_type = 'Alu'\n",
    "binsize = 60\n",
    "nbins = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "allc_path = '/home/AD/rkgadde/L1IP/celltype_allc'\n",
    "genome = '/home/AD/rkgadde/L1IP/hg38_data/hg38.sorted.genome'\n",
    "\n",
    "genedir = '/home/AD/rkgadde/L1IP/gene_data'\n",
    "id_file = f'{genedir}/{me_type}_intersect_mev_ids.txt'\n",
    "\n",
    "me_path = f'/home/AD/rkgadde/L1IP/mC_data/CZI/class/vars'\n",
    "me_files = [f'{me_path}/all_{me_type}_abs.tsv', f'{me_path}/all_{me_type}_ins.tsv']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdir = '/home/AD/rkgadde/L1IP/mC_data/CZI/type'\n",
    "bindir = f'{pdir}/bins'\n",
    "outdir = f'{pdir}/mC'\n",
    "tmpdir = f'{pdir}/tmp'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!grep -f {id_file} {me_files[0]} > {genedir}/{me_type}_intersect_mevs.tsv\n",
    "!grep -f {id_file} {me_files[1]} >> {genedir}/{me_type}_intersect_mevs.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>svlen</th>\n",
       "      <th>strand</th>\n",
       "      <th>mei</th>\n",
       "      <th>het</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17358284</td>\n",
       "      <td>17358591</td>\n",
       "      <td>CZI_abs_6</td>\n",
       "      <td>306.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AluYe5</td>\n",
       "      <td>AF1,AF3,YF1,YM1,YM3</td>\n",
       "      <td>AF2,AM1,AM2,AM3,YF2,YM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24076520</td>\n",
       "      <td>24076812</td>\n",
       "      <td>CZI_abs_7</td>\n",
       "      <td>291.0</td>\n",
       "      <td>-</td>\n",
       "      <td>AluYa5</td>\n",
       "      <td>AM3,YF2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42277515</td>\n",
       "      <td>42277821</td>\n",
       "      <td>CZI_abs_12</td>\n",
       "      <td>305.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AluYa5</td>\n",
       "      <td>AF2,AM3</td>\n",
       "      <td>AF1,AF3,AM2,YF1,YF2,YM1,YM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42542541</td>\n",
       "      <td>42542852</td>\n",
       "      <td>CZI_abs_13</td>\n",
       "      <td>310.0</td>\n",
       "      <td>-</td>\n",
       "      <td>AluY</td>\n",
       "      <td>AF1,AM2,YM1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391798</td>\n",
       "      <td>43392120</td>\n",
       "      <td>CZI_abs_14</td>\n",
       "      <td>321.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AluY</td>\n",
       "      <td>AF1,AM3,YF2</td>\n",
       "      <td>AF2,AM1,AM2,YF1,YM1,YM3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end          id  svlen strand     mei  \\\n",
       "0  chr1  17358284  17358591   CZI_abs_6  306.0      +  AluYe5   \n",
       "1  chr1  24076520  24076812   CZI_abs_7  291.0      -  AluYa5   \n",
       "2  chr1  42277515  42277821  CZI_abs_12  305.0      +  AluYa5   \n",
       "3  chr1  42542541  42542852  CZI_abs_13  310.0      -    AluY   \n",
       "4  chr1  43391798  43392120  CZI_abs_14  321.0      +    AluY   \n",
       "\n",
       "                   het                          alt  \n",
       "0  AF1,AF3,YF1,YM1,YM3      AF2,AM1,AM2,AM3,YF2,YM2  \n",
       "1              AM3,YF2                          NaN  \n",
       "2              AF2,AM3  AF1,AF3,AM2,YF1,YF2,YM1,YM2  \n",
       "3          AF1,AM2,YM1                          NaN  \n",
       "4          AF1,AM3,YF2      AF2,AM1,AM2,YF1,YM1,YM3  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_df = pd.read_csv(f'{genedir}/{me_type}_intersect_mevs.tsv', sep='\\t', \n",
    "                    names=['chrom','start','end','id','svlen','strand','mei','het','alt'])\n",
    "me_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedfile = me_df[['chrom','start','end','id','strand']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bins\n",
    "bed_bins_df = calculate_mC.make_flanking_bins(bedfile, nbins, binsize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Index and merge bins\n",
    "bin_prefix = f'{me_type}_intersect.bins'\n",
    "calculate_mC.process_bins(bed_bins_df, bin_prefix, bindir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate methylation for each bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make directories\n",
    "!mkdir -p {tmpdir}/allc_regions-{binsize}\n",
    "!mkdir -p {tmpdir}/binc-{binsize}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "allc_tables = glob(f'{allc_path}/*.allc.tsv.gz')\n",
    "allc_prefix = f'{tmpdir}/allc_regions-{binsize}/allc_{me_type}_intersect'\n",
    "merge_bed = f'{bindir}/{bin_prefix}.merged.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 231/231 [00:00<00:00, 977683.37it/s]\n"
     ]
    }
   ],
   "source": [
    "# Use tabix to extract the portions of allc tables from the deletion regions.\n",
    "# This will speed up subsequent processing to calculate the binned data\n",
    "with Pool(16) as p:\n",
    "    x = list(tqdm(p.starmap(calculate_mC.tabixallc, \n",
    "                            zip(allc_tables, \n",
    "                                repeat(allc_prefix), \n",
    "                                repeat(merge_bed)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "allcdir = f'{tmpdir}/allc_regions-{binsize}'\n",
    "allc_files = glob(f'{allcdir}/allc_{me_type}_intersect.*.tsv.gz')\n",
    "bed_bins_file = f'{bindir}/{bin_prefix}.bed.gz'\n",
    "binc_prefix = f'{tmpdir}/binc-{binsize}/binc'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample already processed, skipping CGE_LAMP5_LHX6-AF1\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AM2\n",
      "Sample already processed, skipping L2-4IT_CUX2-YM3\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-YM1\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-YM1\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AF2\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-YF1\n",
      "Sample already processed, skipping Glia_Astro-AM3\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-YM1\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-YF1\n",
      "Sample already processed, skipping Glia_Astro-YF2\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-AF3\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-YF2\n",
      "\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AF2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping Glia_Astro-AM1\n",
      "Sample already processed, skipping CGE_PAX6-AF1\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AM2\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AF1\n",
      "Sample already processed, skipping Glia_Oligo-YM3\n",
      "Sample already processed, skipping L2-4IT_CUX2-AF2\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-YF1\n",
      "Sample already processed, skipping Glia_Oligo-AF2\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-YM2\n",
      "Sample already processed, skipping MGE_SST_CLMP-AM2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6-YM2\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AF1\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AM2\n",
      "Sample already processed, skipping CGE_VIP-AF1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_SST-YF1\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-YM1\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-YF2\n",
      "Sample already processed, skipping L6-AM2\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-YM2\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-YM3\n",
      "\n",
      "Sample already processed, skipping Glia_Oligo-YM1\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AF2\n",
      "Sample already processed, skipping CGE_LAMP5-AM2\n",
      "Sample already processed, skipping CGE_PAX6-YM2\n",
      "Sample already processed, skipping MGE_PVALB-AF2\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AM1\n",
      "Sample already processed, skipping CGE_VIP-AM2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_SST_CLMP-YF1\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-YM3\n",
      "Sample already processed, skipping MGE_SST-YM2\n",
      "Sample already processed, skipping Glia_Oligo-YF2\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-YM2\n",
      "\n",
      "Sample already processed, skipping MGE_SST-AF1\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AM3\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AF1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-AF3\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AM1\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-AF1\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-YF2\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-AM2\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AF2\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AF1\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AM2\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping Glia_Astro-AF2\n",
      "\n",
      "Sample already processed, skipping L6-AF3\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-AM1\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AF3\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AF1\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-AM3\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AF3\n",
      "Sample already processed, skipping MGE_SST-AM2\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-AF3\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AF3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AM1\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-AM2\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-YM2\n",
      "Sample already processed, skipping MGE_SST_CLMP-YM2\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-YM2\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AM1\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-YF1\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AM2\n",
      "Sample already processed, skipping MGE_SST_CLMP-AF3\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AF3\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-YF2\n",
      "Sample already processed, skipping Glia_Astro-YM1\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-YF2\n",
      "Sample already processed, skipping CGE_LAMP5-AF3\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-YM1\n",
      "\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-YF2\n",
      "Sample already processed, skipping L6-YF1\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-YF2\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6-AF1\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-YF1\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AF2\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-YM2\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-YF1\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-YM1\n",
      "Sample already processed, skipping Glia_Oligo-AM3\n",
      "Sample already processed, skipping MGE_PVALB-AM1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-YM1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-AM3\n",
      "\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AM3\n",
      "\n",
      "Sample already processed, skipping Glia_Astro-YM3\n",
      "Sample already processed, skipping Glia_Astro-AF3\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-YF1\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-AM2\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-YM1\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AF3\n",
      "Sample already processed, skipping MGE_SST-AF3\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-YM2\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-YM1\n",
      "Sample already processed, skipping MGE_SST_CLMP-AF2\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-YM2\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-AM3\n",
      "Sample already processed, skipping Glia_Oligo-AM2\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-YM3\n",
      "Sample already processed, skipping Glia_Oligo-AM1\n",
      "\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AM2\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-YM2\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-YF2\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-YF1\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AF3\n",
      "Sample already processed, skipping Glia_Astro-YM2\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-YF1\n",
      "Sample already processed, skipping CGE_LAMP5-YF1\n",
      "Sample already processed, skipping L6-AF2\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-YM1\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-YM3\n",
      "Sample already processed, skipping CGE_LAMP5-AF2\n",
      "Sample already processed, skipping MGE_SST-AM1\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_SST_CLMP-AF1\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AF3\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-YM2\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-YF2\n",
      "Sample already processed, skipping L6-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-YF1\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-YF2\n",
      "Sample already processed, skipping MGE_SST_CLMP-YM3\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AM1\n",
      "\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-YF1\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AF2\n",
      "\n",
      "Sample already processed, skipping Glia_Oligo-YF1\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AM2\n",
      "\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AM3\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-YM3\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AF2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AM1\n",
      "Sample already processed, skipping CGE_PAX6-AM3\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-YM1\n",
      "\n",
      "Sample already processed, skipping L6-AM1\n",
      "Sample already processed, skipping CGE_VIP-AM3\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-YF2\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-YF1\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-YF1\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-YM3\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-YF1\n",
      "\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AM3\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6-YM1\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-YM1\n",
      "Sample already processed, skipping Glia_Astro-AF1\n",
      "\n",
      "Sample already processed, skipping Glia_Astro-YF1\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-YM3\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AF2\n",
      "Sample already processed, skipping L2-4IT_CUX2-AM2\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AM1\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_SST-AM3\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AF2\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-AM3\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-AM1\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-YM1\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AF1\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-YM1\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-YF2\n",
      "Sample already processed, skipping CGE_PAX6-AM1\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AF3\n",
      "Sample already processed, skipping MGE_SST-YF2\n",
      "Sample already processed, skipping MGE_PVALB-AF1\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-YM2\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-YF2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-YF1\n",
      "Sample already processed, skipping Glia_Oligo-AF1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AF3\n",
      "Sample already processed, skipping MGE_SST-YM1\n",
      "Sample already processed, skipping L2-4IT_CUX2-AF3\n",
      "Sample already processed, skipping CGE_VIP-AM1\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-YM2\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AF1\n",
      "\n",
      "Sample already processed, skipping L2-4IT_CUX2-AF1\n",
      "Sample already processed, skipping MGE_SST-AF2\n",
      "Sample already processed, skipping MGE_SST_CLMP-YM1\n",
      "\n",
      "Sample already processed, skipping MGE_SST_CLMP-AM1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AM2\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AF3\n",
      "\n",
      "Sample already processed, skipping L6b_TLE4_NXPH4-AF1\n",
      "\n",
      "Sample already processed, skipping MGE_SST-YM3\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AM3\n",
      "Sample already processed, skipping L2-4IT_CUX2-YM2\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-AF1\n",
      "Sample already processed, skipping Glia_Oligo-YM2\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_LRRK1-AM1\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-YM1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-YF2\n",
      "Sample already processed, skipping CGE_PAX6-AF2\n",
      "Sample already processed, skipping Glia_Oligo-AF3\n",
      "Sample already processed, skipping CGE_LAMP5-AM3\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AM3\n",
      "Sample already processed, skipping CGE_ADARB2_ADAM33-YM2\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-AF3\n",
      "\n",
      "Sample already processed, skipping MGE_PVALB-YM2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5-AM1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_LINC00343-YF2\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-AF2\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-YM1\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-AM1\n",
      "Sample already processed, skipping L4-5IT_RORB_TSHZ2-YF2\n",
      "\n",
      "Sample already processed, skipping L4-5IT_RORB_ARHGAP15-AM2\n",
      "\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-AF2\n",
      "\n",
      "\n",
      "Sample already processed, skipping Glia_Astro-AM2\n",
      "\n",
      "\n",
      "Sample already processed, skipping MGE_SST_CLMP-AM3\n",
      "Sample already processed, skipping L3-5IT_RORB_PLCH1-AF1\n",
      "Sample already processed, skipping L6CT_TLE4_FAM95C-YM3\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_VIP-AF2\n",
      "Sample already processed, skipping L6-AM3\n",
      "\n",
      "Sample already processed, skipping CGE_PAX6-YM3\n",
      "\n",
      "Sample already processed, skipping MGE_SST_CLMP-YF2\n",
      "Sample already processed, skipping CGE_VIP-YM3\n",
      "\n",
      "\n",
      "Sample already processed, skipping CGE_LAMP5_LHX6-YM3\n",
      "\n",
      "\n",
      "\n",
      "Sample already processed, skipping L6IT_THEMIS_CUX1-AM1\n",
      "\n",
      "Sample already processed, skipping L56NP_TLE4_TSHZ2-AM2\n",
      "\n",
      "Sample already processed, skipping L6-YF2\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 231/231 [00:00<00:00, 622676.24it/s]\n"
     ]
    }
   ],
   "source": [
    "# Use allcools to calculate the binned mC counts \n",
    "with Pool(16) as p:\n",
    "    x = list(tqdm(p.starmap(calculate_mC.allc2bins, \n",
    "                            zip(allc_files, \n",
    "                                repeat(binc_prefix), \n",
    "                                repeat(f'{me_type}_intersect'), \n",
    "                                repeat(genome), \n",
    "                                repeat(bed_bins_file)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "231"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the binned mC data\n",
    "mc_files = glob(f'{binc_prefix}*{me_type}_intersect_CGN-Both.sparse.bed.gz')\n",
    "samples_celltypes = []\n",
    "for mc_file in mc_files:\n",
    "    sample = re.match(fr'.*binc\\.(.*)\\.{me_type}_intersect_CGN-Both\\.sparse\\.bed\\.gz', mc_file).group(1)\n",
    "    samples_celltypes.append(sample)\n",
    "len(samples_celltypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 231/231 [00:00<00:00, 880003.84it/s]\n"
     ]
    }
   ],
   "source": [
    "# Combine binned mC data\n",
    "with Pool(16) as p:\n",
    "    mc_dfs = list(tqdm(p.starmap(calculate_mC.get_binc, \n",
    "                              zip(samples_celltypes, \n",
    "                                  repeat(binc_prefix), \n",
    "                                  repeat(f'{me_type}_intersect')))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mc_df = pd.concat(mc_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mark genotype for each sample at each locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2882/2882 [00:00<00:00, 1599997.90it/s]\n"
     ]
    }
   ],
   "source": [
    "ids = me_df['id'].unique()\n",
    "\n",
    "with Pool(16) as p:\n",
    "    gt_dfs = list(tqdm(p.starmap(calculate_mC.groupby_gt, \n",
    "                              zip(ids, \n",
    "                                  repeat(me_df), \n",
    "                                  repeat(mc_df), \n",
    "                                  repeat(me_type)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_df = pd.concat(gt_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.51 s, sys: 243 ms, total: 3.75 s\n",
      "Wall time: 4.22 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "filepath = f'{outdir}/binc-{binsize}.{me_type}_intersect.gt.all_samples.parquet.gz'\n",
    "\n",
    "if not os.path.isfile(filepath):\n",
    "    gt_df.to_parquet(filepath, compression='gzip', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
