{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6cd163de-4970-4633-b1a8-9ab89a24cf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from glob import glob\n",
    "from scipy.stats import permutation_test, false_discovery_control, spearmanr\n",
    "import pybedtools\n",
    "import gene_expression\n",
    "\n",
    "from warnings import filterwarnings\n",
    "filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "filterwarnings(\"ignore\", category=pd.errors.SettingWithCopyWarning)\n",
    "\n",
    "#set up paths\n",
    "#=========================\n",
    "workdir = '/home/AD/rkgadde/L1IP'\n",
    "annotation = f'{workdir}/hg38_data/annotations/gencode.v46.basic.annotation.sorted.genes.gff3'\n",
    "celltypes = f'{workdir}/celltypes.txt'\n",
    "\n",
    "genedir = f'{workdir}/gene_data'\n",
    "figdir = f'{workdir}/results/CZI/plots'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9163b55f-80a4-4dc2-9a76-8e4eb311fbdf",
   "metadata": {},
   "source": [
    "# Load in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8079bced-2a7c-49aa-8ca9-d81b160fb29e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cpm</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>4.014924</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>6.679229</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>3.162097</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>4.509863</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000248527</td>\n",
       "      <td>5.159480</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683723</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "      <td>9.264718</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683724</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>11.147137</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683725</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>9.956563</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683726</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>8.790963</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2683727</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>10.021181</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2683728 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample          gene_id        cpm           celltype\n",
       "0          AF1  ENSG00000228463   4.014924  L4-5IT_RORB_TSHZ2\n",
       "1          AF1  ENSG00000230021   6.679229  L4-5IT_RORB_TSHZ2\n",
       "2          AF1  ENSG00000225630   3.162097  L4-5IT_RORB_TSHZ2\n",
       "3          AF1  ENSG00000237973   4.509863  L4-5IT_RORB_TSHZ2\n",
       "4          AF1  ENSG00000248527   5.159480  L4-5IT_RORB_TSHZ2\n",
       "...        ...              ...        ...                ...\n",
       "2683723    YM3  ENSG00000212907   9.264718   L56NP_TLE4_TSHZ2\n",
       "2683724    YM3  ENSG00000198886  11.147137   L56NP_TLE4_TSHZ2\n",
       "2683725    YM3  ENSG00000198786   9.956563   L56NP_TLE4_TSHZ2\n",
       "2683726    YM3  ENSG00000198695   8.790963   L56NP_TLE4_TSHZ2\n",
       "2683727    YM3  ENSG00000198727  10.021181   L56NP_TLE4_TSHZ2\n",
       "\n",
       "[2683728 rows x 4 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load in mes\n",
    "#comb_df = pd.read_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', index_col=0)\n",
    "#load in gene expression\n",
    "cpm_path = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "suffix = '_combinesample_log2cpm+1.tsv.gz'\n",
    "cpm = gene_expression.get_expressed_gene_cpm(cpm_path, suffix, 5, 3) # CPM > 5 in at least 2 samples\n",
    "cpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "259f9cf1-1738-415c-91ed-40a8db88e4d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14651"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cpm[cpm['celltype'] == 'L4-5IT_RORB_TSHZ2']['gene_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bcb723b-cd43-464a-a57f-27d383dc2482",
   "metadata": {},
   "outputs": [],
   "source": [
    "#process mes into joint bed file\n",
    "#process l1\n",
    "me_type = 'L1'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "l1_df = pd.concat([abs_df, ins_df])\n",
    "l1_df['class'] = 'l1'\n",
    "\n",
    "#process alu\n",
    "me_type = 'Alu'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6, 7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "alu_df = pd.concat([abs_df, ins_df])\n",
    "alu_df['class'] = 'alu'\n",
    "\n",
    "#add labels\n",
    "comb_df = pd.concat([l1_df, alu_df])\n",
    "comb_df['start'] = comb_df['start'].astype(int)\n",
    "comb_df['end'] = comb_df['end'].astype(int)\n",
    "comb_df['length'] = comb_df['end'] - comb_df['start']\n",
    "comb_df['het'] = comb_df['het'].fillna('NaN').astype(str)\n",
    "comb_df['hom'] = comb_df['hom'].fillna('NaN').astype(str)\n",
    "comb_df.reset_index(drop=True, inplace=True)\n",
    "het_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['het']])\n",
    "hom_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['hom']])\n",
    "het_count = np.asarray([len(i.split(',')) for i in comb_df['het']])\n",
    "hom_count = np.asarray([len(i.split(',')) for i in comb_df['hom']])\n",
    "ratio = het_count/hom_count\n",
    "#set to NaN and inf\n",
    "ratio[(het_bool == False) & (hom_bool==False)] = np.nan\n",
    "ratio[(het_bool == True) & (hom_bool==False)] = np.inf\n",
    "ratio[(het_bool == False) & (hom_bool==True)] = -1*np.inf\n",
    "comb_df['het_over_hom'] = ratio\n",
    "\n",
    "# label as majority het or hom, if >2x\n",
    "genotype = np.empty(len(comb_df)).astype(str)\n",
    "genotype[ratio >= 2] = 'het'\n",
    "genotype[ratio <= 0.5] = 'hom'\n",
    "genotype[(ratio <= 2) & (ratio >= 0.5)] = 'mixed'\n",
    "genotype[np.isnan(ratio)] = 'NaN'\n",
    "comb_df['genotype'] = genotype\n",
    "#label as truncated or full length\n",
    "trunc = np.empty(len(comb_df), dtype=object)\n",
    "l1_mask = comb_df['class'] == 'l1'\n",
    "trunc[l1_mask & (comb_df['length'] > 5500)] = 'full_length'\n",
    "trunc[l1_mask & (comb_df['length'] <= 5500)] = 'truncated'\n",
    "alu_mask = comb_df['class'] == 'alu'\n",
    "trunc[alu_mask & (comb_df['length'] > 280)] = 'full_length'\n",
    "trunc[alu_mask & (comb_df['length'] <= 280)] = 'truncated'\n",
    "comb_df['insertion_category'] = trunc\n",
    "\n",
    "comb_df.to_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef9501d-61f3-4a14-a826-84eb828a2645",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#Get bed file with gene names\n",
    "#==================================\n",
    "cd /cndd3/dburrows/DATA/annotations/gencode/\n",
    "#Split up 4th column into gene id and type\n",
    "# awk 'BEGIN{OFS=\"\\t\"} {split($4, a, \"_\"); $4 = a[1]; $5 = a[2]; print}' gencode.v44.annotation.hg38.genelabels.bed > tmp_split.bed\n",
    "# mv tmp_split.bed gencode.v44.annotation.hg38.genelabels.bed\n",
    "\n",
    "#remove duplicated gene loci (exons from isoforms) that have the same chromosome, start+end position gene_label, and strand\n",
    "# awk '!seen[$1,$2,$3,$5,$6]++' gencode.v44.annotation.hg38.genelabels.bed > tmp_seen.bed\n",
    "# mv tmp_seen.bed gencode.v44.annotation.hg38.genelabels.unique.bed\n",
    "awk -F'\\t' '{\n",
    "    # Extract necessary fields from the 9th column (attributes)\n",
    "    match($9, /gene_id=([^;]+)/, gene_id);\n",
    "    match($9, /gene_name=([^;]+)/, gene_name);\n",
    "    match($9, /gene_type=([^;]+)/, gene_type);\n",
    "\n",
    "    # Rearrange and print the columns\n",
    "    print $1, $4, $5, $2, $3, $7, gene_id[1], gene_name[1], gene_type[1];\n",
    "}' OFS=\"\\t\" gencode.v46.basic.annotation.sorted.genes.gff3 > gencode.v46.basic.annotation.sorted.genes.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cc721cc9-973a-4396-9c54-9d6fd2a7c0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection of mes with genes\n",
    "#bedconvert and load annotations\n",
    "me_bt = pybedtools.BedTool.from_dataframe(comb_df)\n",
    "gene_df = pd.read_csv('/cndd3/dburrows/DATA/annotations/gencode/gencode.v46.basic.annotation.sorted.genes.bed', sep='\\t',header=None)\n",
    "gene_bt = pybedtools.BedTool.from_dataframe(gene_df)\n",
    "\n",
    "#compute intersection\n",
    "me_gene = gene_bt.intersect(me_bt, f=0.01, wo=True) \n",
    "me_gene = me_gene.to_dataframe(disable_auto_names=True, header=None)\n",
    "me_gene.rename(columns={0: 'chr', 1: 'gene_start', 2: 'gene_end', 5: 'gene_strand',\n",
    "                        6: 'gene_id', 7: 'gene_name', 8: 'gene_type', 10: 'me_start',\n",
    "                       11: 'me_end', 12: 'me_id', 14: 'me_strand', 15: 'me_class', 16: 'het',\n",
    "                       17: 'hom', 18: 'me_type', 19: 'het_hom_ratio', 20:'genotype', \n",
    "                       21: 'insertion_category', 22: 'length'}, inplace=True)\n",
    "me_gene.drop(columns=np.asarray(me_gene.columns[(np.asarray([type(i) for i in me_gene.columns]) == int)]), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1918d864-9099-4863-863f-c921e9419b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection with genomic regions\n",
    "\n",
    "#compute intersection\n",
    "l1_cgi = l1_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_cgi = alu_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_all = alu_bt.intersect(all_bt, f=0.3, wo=True) \n",
    "l1_all = all_bt.intersect(l1_bt, f=0.3, wo=True) \n",
    "\n",
    "l1_cgi_df = l1_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_cgi_df = alu_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_all_df = alu_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_all_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "l1_all_df = l1_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_all_df.rename(columns={3: 'region', 15: 'me_type', 17: 'genotype', 18: 'insertion_category', 13: 'het', 14: 'hom', 10: 'length'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3735c321-84a2-4c18-a945-e1ca569e2b23",
   "metadata": {},
   "source": [
    "# eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3822f5c1-aec2-44eb-b2fd-7000e8e9756b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_over_hom</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>3838</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65443954</td>\n",
       "      <td>65444209</td>\n",
       "      <td>CZI_abs_1545</td>\n",
       "      <td>255</td>\n",
       "      <td>-</td>\n",
       "      <td>l1</td>\n",
       "      <td>YM2</td>\n",
       "      <td>AF1,YM1,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>80939086</td>\n",
       "      <td>80945263</td>\n",
       "      <td>CZI_abs_1546</td>\n",
       "      <td>6177</td>\n",
       "      <td>-</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF2,AF3,AM1,YF1,YM1,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>82660292</td>\n",
       "      <td>82661886</td>\n",
       "      <td>CZI_abs_1547</td>\n",
       "      <td>1594</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AF2,AM2,YF1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>86275302</td>\n",
       "      <td>86276014</td>\n",
       "      <td>CZI_abs_1548</td>\n",
       "      <td>712</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124088197</td>\n",
       "      <td>124088200</td>\n",
       "      <td>CZI_ins_4331</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF3,AM1,YF1,YF2,YM2</td>\n",
       "      <td>AF2,AM2,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>mixed</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>chr9</td>\n",
       "      <td>126217880</td>\n",
       "      <td>126217897</td>\n",
       "      <td>CZI_ins_4333</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AM1,YM1</td>\n",
       "      <td>AF1,AF2,AF3,YF2,YM2</td>\n",
       "      <td>insertion</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>chr9</td>\n",
       "      <td>127834834</td>\n",
       "      <td>127834845</td>\n",
       "      <td>CZI_ins_4334</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM1,YM2</td>\n",
       "      <td>YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>het</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129426079</td>\n",
       "      <td>129426094</td>\n",
       "      <td>CZI_ins_4335</td>\n",
       "      <td>15</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000</th>\n",
       "      <td>chr9</td>\n",
       "      <td>131003481</td>\n",
       "      <td>131003497</td>\n",
       "      <td>CZI_ins_4336</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>AF3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>het</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5001 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end            id  length strand class  \\\n",
       "0     chr1   56365451   56369289  CZI_abs_1543    3838      +    l1   \n",
       "1     chr1   65443954   65444209  CZI_abs_1545     255      -    l1   \n",
       "2     chr1   80939086   80945263  CZI_abs_1546    6177      -    l1   \n",
       "3     chr1   82660292   82661886  CZI_abs_1547    1594      +    l1   \n",
       "4     chr1   86275302   86276014  CZI_abs_1548     712      +    l1   \n",
       "...    ...        ...        ...           ...     ...    ...   ...   \n",
       "4996  chr9  124088197  124088200  CZI_ins_4331       3      -   alu   \n",
       "4997  chr9  126217880  126217897  CZI_ins_4333      17      +   alu   \n",
       "4998  chr9  127834834  127834845  CZI_ins_4334      11      -   alu   \n",
       "4999  chr9  129426079  129426094  CZI_ins_4335      15      -   alu   \n",
       "5000  chr9  131003481  131003497  CZI_ins_4336      16      +   alu   \n",
       "\n",
       "                                          het  \\\n",
       "0                                         AF3   \n",
       "1                                         YM2   \n",
       "2                                         NaN   \n",
       "3                                         NaN   \n",
       "4                                         NaN   \n",
       "...                                       ...   \n",
       "4996                  AF1,AF3,AM1,YF1,YF2,YM2   \n",
       "4997                                  AM1,YM1   \n",
       "4998      AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM1,YM2   \n",
       "4999                                      NaN   \n",
       "5000  AF1,AF2,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3   \n",
       "\n",
       "                                              hom    me_type  het_over_hom  \\\n",
       "0                             AM1,AM2,AM3,YF2,YM3    absence      0.200000   \n",
       "1                                     AF1,YM1,YM3    absence      0.333333   \n",
       "2                         AF2,AF3,AM1,YF1,YM1,YM3    absence          -inf   \n",
       "3                                 AF1,AF2,AM2,YF1    absence          -inf   \n",
       "4                         AF1,AM1,YF1,YM1,YM2,YM3    absence          -inf   \n",
       "...                                           ...        ...           ...   \n",
       "4996                                  AF2,AM2,YM3  insertion      2.000000   \n",
       "4997                          AF1,AF2,AF3,YF2,YM2  insertion      0.400000   \n",
       "4998                                          YM3  insertion      9.000000   \n",
       "4999  AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3  insertion          -inf   \n",
       "5000                                          AF3  insertion     10.000000   \n",
       "\n",
       "     genotype insertion_category  \n",
       "0         hom          truncated  \n",
       "1         hom          truncated  \n",
       "2         hom        full_length  \n",
       "3         hom          truncated  \n",
       "4         hom          truncated  \n",
       "...       ...                ...  \n",
       "4996    mixed          truncated  \n",
       "4997      hom          truncated  \n",
       "4998      het          truncated  \n",
       "4999      hom          truncated  \n",
       "5000      het          truncated  \n",
       "\n",
       "[5001 rows x 13 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fc59d133-fff3-477c-aea6-a4774b60798b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24066774</td>\n",
       "      <td>24083565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230703.1</td>\n",
       "      <td>MYOM3-AS1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>24076520</td>\n",
       "      <td>24076812</td>\n",
       "      <td>CZI_abs_7</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>AM3,YF2</td>\n",
       "      <td>absence</td>\n",
       "      <td>4.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55915603</td>\n",
       "      <td>55944980</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230250.3</td>\n",
       "      <td>LINC01753</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>55935619</td>\n",
       "      <td>55935930</td>\n",
       "      <td>CZI_abs_19</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56154545</td>\n",
       "      <td>56477687</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260971.4</td>\n",
       "      <td>ENSG00000260971</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56173433</td>\n",
       "      <td>56524495</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284686.1</td>\n",
       "      <td>ENSG00000284686</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57605847</td>\n",
       "      <td>57606206</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000236888.1</td>\n",
       "      <td>RPS20P5</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>57606184</td>\n",
       "      <td>57606197</td>\n",
       "      <td>CZI_ins_818</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AM3,YF1</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,YM1,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>0.2857142857142857</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>chr8</td>\n",
       "      <td>143636019</td>\n",
       "      <td>143656418</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000183309.12</td>\n",
       "      <td>ZNF623</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143641445</td>\n",
       "      <td>143641764</td>\n",
       "      <td>CZI_abs_1430</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF2,AM3,YM1,YM2</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>chr8</td>\n",
       "      <td>144772224</td>\n",
       "      <td>144787348</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000196378.14</td>\n",
       "      <td>ZNF34</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>144776325</td>\n",
       "      <td>144776913</td>\n",
       "      <td>CZI_abs_1433</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>9.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>chr9</td>\n",
       "      <td>35791003</td>\n",
       "      <td>35809732</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159899.16</td>\n",
       "      <td>NPR2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>35803110</td>\n",
       "      <td>35803402</td>\n",
       "      <td>CZI_abs_1464</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AM1,AM2,YF1,YF2,YM2,YM3</td>\n",
       "      <td>AF2,AF3,AM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>2.333333333333333</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>chr9</td>\n",
       "      <td>73042745</td>\n",
       "      <td>73056528</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000228024.1</td>\n",
       "      <td>CYP1D1P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>73054613</td>\n",
       "      <td>73054880</td>\n",
       "      <td>CZI_abs_1473</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129430652</td>\n",
       "      <td>129451422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230676.1</td>\n",
       "      <td>ENSG00000230676</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129435084</td>\n",
       "      <td>129435407</td>\n",
       "      <td>CZI_abs_1499</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr  gene_start   gene_end gene_strand             gene_id  \\\n",
       "0    chr1    24066774   24083565           +   ENSG00000230703.1   \n",
       "1    chr1    55915603   55944980           -   ENSG00000230250.3   \n",
       "2    chr1    56154545   56477687           -   ENSG00000260971.4   \n",
       "3    chr1    56173433   56524495           -   ENSG00000284686.1   \n",
       "4    chr1    57605847   57606206           -   ENSG00000236888.1   \n",
       "..    ...         ...        ...         ...                 ...   \n",
       "122  chr8   143636019  143656418           +  ENSG00000183309.12   \n",
       "123  chr8   144772224  144787348           -  ENSG00000196378.14   \n",
       "124  chr9    35791003   35809732           +  ENSG00000159899.16   \n",
       "125  chr9    73042745   73056528           +   ENSG00000228024.1   \n",
       "126  chr9   129430652  129451422           -   ENSG00000230676.1   \n",
       "\n",
       "           gene_name               gene_type   me_start     me_end  \\\n",
       "0          MYOM3-AS1                  lncRNA   24076520   24076812   \n",
       "1          LINC01753                  lncRNA   55935619   55935930   \n",
       "2    ENSG00000260971                  lncRNA   56365451   56369289   \n",
       "3    ENSG00000284686          protein_coding   56365451   56369289   \n",
       "4            RPS20P5    processed_pseudogene   57606184   57606197   \n",
       "..               ...                     ...        ...        ...   \n",
       "122           ZNF623          protein_coding  143641445  143641764   \n",
       "123            ZNF34          protein_coding  144776325  144776913   \n",
       "124             NPR2          protein_coding   35803110   35803402   \n",
       "125          CYP1D1P  unprocessed_pseudogene   73054613   73054880   \n",
       "126  ENSG00000230676                  lncRNA  129435084  129435407   \n",
       "\n",
       "            me_id me_strand me_class                                  het  \\\n",
       "0       CZI_abs_7         -      alu      AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3   \n",
       "1      CZI_abs_19         +      alu                                    .   \n",
       "2    CZI_abs_1543         +       l1                                  AF3   \n",
       "3    CZI_abs_1543         +       l1                                  AF3   \n",
       "4     CZI_ins_818         -      alu                              AM3,YF1   \n",
       "..            ...       ...      ...                                  ...   \n",
       "122  CZI_abs_1430         -      alu                                    .   \n",
       "123  CZI_abs_1433         +      alu  AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3   \n",
       "124  CZI_abs_1464         -      alu          AF1,AM1,AM2,YF1,YF2,YM2,YM3   \n",
       "125  CZI_abs_1473         +      alu                                    .   \n",
       "126  CZI_abs_1499         -      alu                                    .   \n",
       "\n",
       "                                             hom    me_type  \\\n",
       "0                                        AM3,YF2    absence   \n",
       "1                                            AF1    absence   \n",
       "2                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "3                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "4                    AF1,AF2,AF3,AM1,AM2,YM1,YM3  insertion   \n",
       "..                                           ...        ...   \n",
       "122                              AF2,AM3,YM1,YM2    absence   \n",
       "123                                          YM1    absence   \n",
       "124                                  AF2,AF3,AM3    absence   \n",
       "125                                          YM1    absence   \n",
       "126  AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3    absence   \n",
       "\n",
       "          het_hom_ratio genotype insertion_category  length  \n",
       "0                   4.0      het        full_length     292  \n",
       "1                  -inf      hom        full_length     311  \n",
       "2                   0.2      hom          truncated    3838  \n",
       "3                   0.2      hom          truncated    3838  \n",
       "4    0.2857142857142857      hom          truncated      13  \n",
       "..                  ...      ...                ...     ...  \n",
       "122                -inf      hom        full_length     319  \n",
       "123                 9.0      het        full_length     588  \n",
       "124   2.333333333333333      het        full_length     292  \n",
       "125                -inf      hom          truncated     267  \n",
       "126                -inf      hom        full_length     323  \n",
       "\n",
       "[127 rows x 19 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_gene[]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c28620cf-3a14-49a2-bd54-e4a718ed3a95",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset to directly overlapping genes first\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df8a112b-9e0c-4a26-8b14-bb5f9c5798f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#write code to test association with each gene, including covariates, linreg\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7aff4fe-670a-4487-8cd9-e16280cef709",
   "metadata": {},
   "outputs": [],
   "source": [
    "#select genes to test, within certain range, reduce FDR, run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27a2c7fc-3bce-40fb-b959-c1949b0c0a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualise significant effects, by celltype, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2938ba-9163-4cd1-a42c-7c2b2343e46c",
   "metadata": {},
   "source": [
    "# metheQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ef7a6b-16ef-432d-ac80-cd038a611adf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
