{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6cd163de-4970-4633-b1a8-9ab89a24cf3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from glob import glob\n",
    "from scipy.stats import permutation_test, false_discovery_control, spearmanr\n",
    "import pybedtools\n",
    "import gene_expression\n",
    "import process as pfn\n",
    "\n",
    "from warnings import filterwarnings\n",
    "filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "filterwarnings(\"ignore\", category=pd.errors.SettingWithCopyWarning)\n",
    "%load_ext autoreload\n",
    "#set up paths\n",
    "#=========================\n",
    "workdir = '/home/AD/rkgadde/L1IP'\n",
    "annotation = f'{workdir}/hg38_data/annotations/gencode.v46.basic.annotation.sorted.genes.gff3'\n",
    "celltypes = f'{workdir}/celltypes.txt'\n",
    "\n",
    "genedir = f'{workdir}/gene_data'\n",
    "figdir = f'{workdir}/results/CZI/plots'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9163b55f-80a4-4dc2-9a76-8e4eb311fbdf",
   "metadata": {},
   "source": [
    "# Load in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8079bced-2a7c-49aa-8ca9-d81b160fb29e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in mes\n",
    "comb_df = pd.read_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', index_col=0)\n",
    "samples = pd.read_csv(f'{workdir}/samples.txt', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bcb723b-cd43-464a-a57f-27d383dc2482",
   "metadata": {},
   "outputs": [],
   "source": [
    "#process mes into joint bed file\n",
    "#process l1\n",
    "me_type = 'L1'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "l1_df = pd.concat([abs_df, ins_df])\n",
    "l1_df['class'] = 'l1'\n",
    "\n",
    "#process alu\n",
    "me_type = 'Alu'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6, 7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "alu_df = pd.concat([abs_df, ins_df])\n",
    "alu_df['class'] = 'alu'\n",
    "\n",
    "#add labels\n",
    "comb_df = pd.concat([l1_df, alu_df])\n",
    "comb_df['start'] = comb_df['start'].astype(int)\n",
    "comb_df['end'] = comb_df['end'].astype(int)\n",
    "comb_df['length'] = comb_df['end'] - comb_df['start']\n",
    "comb_df['het'] = comb_df['het'].fillna('NaN').astype(str)\n",
    "comb_df['hom'] = comb_df['hom'].fillna('NaN').astype(str)\n",
    "comb_df.reset_index(drop=True, inplace=True)\n",
    "het_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['het']])\n",
    "hom_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['hom']])\n",
    "het_count = np.asarray([len(i.split(',')) for i in comb_df['het']])\n",
    "hom_count = np.asarray([len(i.split(',')) for i in comb_df['hom']])\n",
    "ratio = het_count/hom_count\n",
    "#set to NaN and inf\n",
    "ratio[(het_bool == False) & (hom_bool==False)] = np.nan\n",
    "ratio[(het_bool == True) & (hom_bool==False)] = np.inf\n",
    "ratio[(het_bool == False) & (hom_bool==True)] = -1*np.inf\n",
    "comb_df['het_over_hom'] = ratio\n",
    "\n",
    "# label as majority het or hom, if >2x\n",
    "genotype = np.empty(len(comb_df)).astype(str)\n",
    "genotype[ratio >= 2] = 'het'\n",
    "genotype[ratio <= 0.5] = 'hom'\n",
    "genotype[(ratio <= 2) & (ratio >= 0.5)] = 'mixed'\n",
    "genotype[np.isnan(ratio)] = 'NaN'\n",
    "comb_df['genotype'] = genotype\n",
    "#label as truncated or full length\n",
    "trunc = np.empty(len(comb_df), dtype=object)\n",
    "l1_mask = comb_df['class'] == 'l1'\n",
    "trunc[l1_mask & (comb_df['length'] > 5500)] = 'full_length'\n",
    "trunc[l1_mask & (comb_df['length'] <= 5500)] = 'truncated'\n",
    "alu_mask = comb_df['class'] == 'alu'\n",
    "trunc[alu_mask & (comb_df['length'] > 280)] = 'full_length'\n",
    "trunc[alu_mask & (comb_df['length'] <= 280)] = 'truncated'\n",
    "comb_df['insertion_category'] = trunc\n",
    "\n",
    "comb_df.to_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef9501d-61f3-4a14-a826-84eb828a2645",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#Get bed file with gene names\n",
    "#==================================\n",
    "cd /cndd3/dburrows/DATA/annotations/gencode/\n",
    "#Split up 4th column into gene id and type\n",
    "# awk 'BEGIN{OFS=\"\\t\"} {split($4, a, \"_\"); $4 = a[1]; $5 = a[2]; print}' gencode.v44.annotation.hg38.genelabels.bed > tmp_split.bed\n",
    "# mv tmp_split.bed gencode.v44.annotation.hg38.genelabels.bed\n",
    "\n",
    "#remove duplicated gene loci (exons from isoforms) that have the same chromosome, start+end position gene_label, and strand\n",
    "# awk '!seen[$1,$2,$3,$5,$6]++' gencode.v44.annotation.hg38.genelabels.bed > tmp_seen.bed\n",
    "# mv tmp_seen.bed gencode.v44.annotation.hg38.genelabels.unique.bed\n",
    "awk -F'\\t' '{\n",
    "    # Extract necessary fields from the 9th column (attributes)\n",
    "    match($9, /gene_id=([^;]+)/, gene_id);\n",
    "    match($9, /gene_name=([^;]+)/, gene_name);\n",
    "    match($9, /gene_type=([^;]+)/, gene_type);\n",
    "\n",
    "    # Rearrange and print the columns\n",
    "    print $1, $4, $5, $2, $3, $7, gene_id[1], gene_name[1], gene_type[1];\n",
    "}' OFS=\"\\t\" gencode.v46.basic.annotation.sorted.genes.gff3 > gencode.v46.basic.annotation.sorted.genes.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc721cc9-973a-4396-9c54-9d6fd2a7c0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection of mes with genes\n",
    "#bedconvert and load annotations\n",
    "me_bt = pybedtools.BedTool.from_dataframe(comb_df)\n",
    "gene_df = pd.read_csv('/cndd3/dburrows/DATA/annotations/gencode/gencode.v46.basic.annotation.sorted.genes.bed', sep='\\t',header=None)\n",
    "gene_bt = pybedtools.BedTool.from_dataframe(gene_df)\n",
    "\n",
    "#compute intersection\n",
    "me_gene = gene_bt.intersect(me_bt, f=0.01, wo=True) \n",
    "me_gene = me_gene.to_dataframe(disable_auto_names=True, header=None)\n",
    "me_gene.rename(columns={0: 'chr', 1: 'gene_start', 2: 'gene_end', 5: 'gene_strand',\n",
    "                        6: 'gene_id', 7: 'gene_name', 8: 'gene_type', 10: 'me_start',\n",
    "                       11: 'me_end', 12: 'me_id', 14: 'me_strand', 15: 'me_class', 16: 'het',\n",
    "                       17: 'hom', 18: 'me_type', 19: 'het_hom_ratio', 20:'genotype', \n",
    "                       21: 'insertion_category', 22: 'length'}, inplace=True)\n",
    "me_gene.drop(columns=np.asarray(me_gene.columns[(np.asarray([type(i) for i in me_gene.columns]) == int)]), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1918d864-9099-4863-863f-c921e9419b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection with genomic regions\n",
    "\n",
    "#compute intersection\n",
    "l1_cgi = l1_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_cgi = alu_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_all = alu_bt.intersect(all_bt, f=0.3, wo=True) \n",
    "l1_all = all_bt.intersect(l1_bt, f=0.3, wo=True) \n",
    "\n",
    "l1_cgi_df = l1_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_cgi_df = alu_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_all_df = alu_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_all_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "l1_all_df = l1_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_all_df.rename(columns={3: 'region', 15: 'me_type', 17: 'genotype', 18: 'insertion_category', 13: 'het', 14: 'hom', 10: 'length'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3735c321-84a2-4c18-a945-e1ca569e2b23",
   "metadata": {},
   "source": [
    "# eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "1bb6af47-e48c-4fcb-b2da-a92924dce8ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████| 127/127 [02:03<00:00,  1.03it/s]\n",
      "/tmp/ipykernel_1744016/1928928272.py:64: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  sig_df = sig_df.groupby('celltype').apply(fdr).reset_index()\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "import time\n",
    "#run for MEs directly overlapping genes\n",
    "#=====================================\n",
    "\n",
    "#load in + filter cpm data\n",
    "me_gene['gene_id'] = np.asarray([i.split('.')[0] for i in me_gene['gene_id']])\n",
    "cpm_path = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "cpm = pfn.combine_genes_bycell_filter(rna = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , mini = 0\n",
    "                                      , n_samp = 0)\n",
    "\n",
    "#extract sample and gene values\n",
    "samples = cpm['sample'].unique()\n",
    "cpm_unq = cpm['gene_id'].unique()\n",
    "\n",
    "out_df = pd.DataFrame()\n",
    "it=range(len(me_gene))\n",
    "#iterate over each gene\n",
    "#============================\n",
    "for i in tqdm(it, leave=True):\n",
    "    time.sleep(0.1)\n",
    "    curr = pd.DataFrame(me_gene.iloc[i]).T\n",
    "    cpm_gene = cpm[cpm['gene_id'] == curr['gene_id'].values[0]]\n",
    "    #skip genes that are not included in the filtered list\n",
    "    if len(cpm_gene) > 0:\n",
    "        assert len(cpm_gene) == len(cpm_gene['celltype'].unique())*len(samples), 'some samples missing' #check that the correct n samples are there\n",
    "    \n",
    "        #define design matrix for model\n",
    "        #============================\n",
    "        cpm_gene.set_index(['celltype', 'sample'], inplace=True)\n",
    "        cpm_gene.drop(columns='gene_id', inplace=True)\n",
    "        cpm_gene['age'] = [i[0] for i in cpm_gene.index.get_level_values('sample')]\n",
    "        cpm_gene['sex'] = [i[1] for i in cpm_gene.index.get_level_values('sample')]\n",
    "        cpm_gene['age_encoded'] = cpm_gene['age'].map({'Y': 0, 'A': 1})\n",
    "        cpm_gene['sex_encoded'] = cpm_gene['sex'].map({'M': 0, 'F': 1})\n",
    "        \n",
    "        #define me_type values, 0,1,2\n",
    "        if curr['me_type'].values == 'insertion': \n",
    "            hom_val = 2\n",
    "            wt_val = 0\n",
    "        if curr['me_type'].values == 'absence':\n",
    "            hom_val = 0\n",
    "            wt_val = 2\n",
    "        vec = np.full(len(cpm_gene), np.nan)\n",
    "        vec[cpm_gene.index.get_level_values('sample').str.contains(curr['het'].values[0].replace(',','|'))] = 1\n",
    "        vec[cpm_gene.index.get_level_values('sample').str.contains(curr['hom'].values[0].replace(',','|'))] = hom_val\n",
    "        wt = ','.join(samples[np.in1d(samples, np.append(curr['het'].values[0].split(','), \n",
    "                                                        curr['hom'].values[0].split(','))) == False]).replace(',','|')\n",
    "        if len(wt) > 0: vec[cpm_gene.index.get_level_values('sample').str.contains(wt)] = wt_val\n",
    "        cpm_gene['me_type'] = vec\n",
    "    \n",
    "        #run model\n",
    "        #==========\n",
    "        out = cpm_gene.groupby('celltype').apply(pfn.run_regression).reset_index()\n",
    "        out['contrast'] = f\"{curr['me_id'].values[0]}_{curr['gene_id'].values[0]}\"\n",
    "        out_df = pd.concat([out_df, out])\n",
    "\n",
    "#get FDR corrected p values\n",
    "group = out_df[out_df['level_1']=='me_type']\n",
    "nanbool = np.isnan(group['p_values'])\n",
    "group = group[~nanbool]\n",
    "pv = group['p_values']\n",
    "reject, pvals_corrected, _, _ = multipletests(pv, alpha=0.05, method='fdr_bh')\n",
    "results = pd.DataFrame({\n",
    "    'coefficients': group['coefficients'],\n",
    "    'p_values': group['p_values'],\n",
    "    'contrast' : group['contrast'],\n",
    "    'padj': pvals_corrected\n",
    "})\n",
    "#get FDR corrected p values\n",
    "#sig_df = sig_df.groupby('celltype').apply(fdr).reset_index()\n",
    "group = out_df[out_df['level_1']=='me_type']\n",
    "nanbool = np.isnan(group['p_values'])\n",
    "group = group[~nanbool]\n",
    "pv = group['p_values']\n",
    "reject, pvals_corrected, _, _ = multipletests(pv, alpha=0.05, method='fdr_bh')\n",
    "group['padj'] = pvals_corrected\n",
    "sig_df = group[group['padj'] < 0.01].reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "6061d6f7-a6ee-4950-a5f5-0a93c76325d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>level_1</th>\n",
       "      <th>coefficients</th>\n",
       "      <th>p_values</th>\n",
       "      <th>contrast</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGE_ADARB2_ADAM33</td>\n",
       "      <td>me_type</td>\n",
       "      <td>0.856286</td>\n",
       "      <td>6.541094e-06</td>\n",
       "      <td>CZI_abs_64_ENSG00000143452</td>\n",
       "      <td>0.000352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L6IT_THEMIS_LINC00343</td>\n",
       "      <td>me_type</td>\n",
       "      <td>0.872579</td>\n",
       "      <td>2.418175e-04</td>\n",
       "      <td>CZI_abs_64_ENSG00000143452</td>\n",
       "      <td>0.008542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L6IT_THEMIS_LINC00343</td>\n",
       "      <td>me_type</td>\n",
       "      <td>-0.114944</td>\n",
       "      <td>2.653709e-04</td>\n",
       "      <td>CZI_abs_66_ENSG00000226949</td>\n",
       "      <td>0.009238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L3-5IT_RORB_PLCH1</td>\n",
       "      <td>me_type</td>\n",
       "      <td>-0.835541</td>\n",
       "      <td>2.408572e-04</td>\n",
       "      <td>CZI_abs_193_ENSG00000203258</td>\n",
       "      <td>0.008542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGE_LAMP5</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.064408</td>\n",
       "      <td>6.466298e-05</td>\n",
       "      <td>CZI_abs_302_ENSG00000110801</td>\n",
       "      <td>0.002589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>L6IT_THEMIS_LINC00343</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.073876</td>\n",
       "      <td>4.136368e-07</td>\n",
       "      <td>CZI_abs_1301_ENSG00000135185</td>\n",
       "      <td>0.000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>L6b_TLE4_NXPH4</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.059364</td>\n",
       "      <td>9.978457e-07</td>\n",
       "      <td>CZI_abs_1301_ENSG00000135185</td>\n",
       "      <td>0.000085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MGE_PVALB</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.252902</td>\n",
       "      <td>4.117493e-07</td>\n",
       "      <td>CZI_abs_1301_ENSG00000135185</td>\n",
       "      <td>0.000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>MGE_SST</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.145040</td>\n",
       "      <td>1.223470e-08</td>\n",
       "      <td>CZI_abs_1301_ENSG00000135185</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MGE_SST_CLMP</td>\n",
       "      <td>me_type</td>\n",
       "      <td>2.302827</td>\n",
       "      <td>9.922971e-06</td>\n",
       "      <td>CZI_abs_1301_ENSG00000135185</td>\n",
       "      <td>0.000486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 celltype  level_1  coefficients      p_values  \\\n",
       "0       CGE_ADARB2_ADAM33  me_type      0.856286  6.541094e-06   \n",
       "1   L6IT_THEMIS_LINC00343  me_type      0.872579  2.418175e-04   \n",
       "2   L6IT_THEMIS_LINC00343  me_type     -0.114944  2.653709e-04   \n",
       "3       L3-5IT_RORB_PLCH1  me_type     -0.835541  2.408572e-04   \n",
       "4               CGE_LAMP5  me_type      2.064408  6.466298e-05   \n",
       "..                    ...      ...           ...           ...   \n",
       "65  L6IT_THEMIS_LINC00343  me_type      2.073876  4.136368e-07   \n",
       "66         L6b_TLE4_NXPH4  me_type      2.059364  9.978457e-07   \n",
       "67              MGE_PVALB  me_type      2.252902  4.117493e-07   \n",
       "68                MGE_SST  me_type      2.145040  1.223470e-08   \n",
       "69           MGE_SST_CLMP  me_type      2.302827  9.922971e-06   \n",
       "\n",
       "                        contrast      padj  \n",
       "0     CZI_abs_64_ENSG00000143452  0.000352  \n",
       "1     CZI_abs_64_ENSG00000143452  0.008542  \n",
       "2     CZI_abs_66_ENSG00000226949  0.009238  \n",
       "3    CZI_abs_193_ENSG00000203258  0.008542  \n",
       "4    CZI_abs_302_ENSG00000110801  0.002589  \n",
       "..                           ...       ...  \n",
       "65  CZI_abs_1301_ENSG00000135185  0.000050  \n",
       "66  CZI_abs_1301_ENSG00000135185  0.000085  \n",
       "67  CZI_abs_1301_ENSG00000135185  0.000050  \n",
       "68  CZI_abs_1301_ENSG00000135185  0.000005  \n",
       "69  CZI_abs_1301_ENSG00000135185  0.000486  \n",
       "\n",
       "[70 rows x 6 columns]"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "bdcf1c45-f867-46d1-8747-7dc0e59522ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cpm</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_encoded</th>\n",
       "      <th>sex_encoded</th>\n",
       "      <th>me_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L4-5IT_RORB_TSHZ2</th>\n",
       "      <th>AF1</th>\n",
       "      <td>0.383286</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF2</th>\n",
       "      <td>1.516757</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF3</th>\n",
       "      <td>0.376111</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM1</th>\n",
       "      <td>0.783174</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF1</th>\n",
       "      <td>0.273898</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L56NP_TLE4_TSHZ2</th>\n",
       "      <th>YF1</th>\n",
       "      <td>0.378245</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF2</th>\n",
       "      <td>0.427399</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM1</th>\n",
       "      <td>0.481128</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM2</th>\n",
       "      <td>1.584667</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM3</th>\n",
       "      <td>0.174273</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cpm age sex  age_encoded  sex_encoded  me_type\n",
       "celltype          sample                                                     \n",
       "L4-5IT_RORB_TSHZ2 AF1     0.383286   A   F            1            1      0.0\n",
       "                  AF2     1.516757   A   F            1            1      0.0\n",
       "                  AF3     0.376111   A   F            1            1      0.0\n",
       "                  AM1     0.783174   A   M            1            0      0.0\n",
       "                  YF1     0.273898   Y   F            0            1      0.0\n",
       "...                            ...  ..  ..          ...          ...      ...\n",
       "L56NP_TLE4_TSHZ2  YF1     0.378245   Y   F            0            1      0.0\n",
       "                  YF2     0.427399   Y   F            0            1      0.0\n",
       "                  YM1     0.481128   Y   M            0            0      0.0\n",
       "                  YM2     1.584667   Y   M            0            0      0.0\n",
       "                  YM3     0.174273   Y   M            0            0      0.0\n",
       "\n",
       "[180 rows x 6 columns]"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "id": "58be0f06-100a-4ba7-994e-3499337159cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cpm</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2436</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>0.556982</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48651</th>\n",
       "      <td>AF2</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>1.850423</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94866</th>\n",
       "      <td>AF3</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>0.565452</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141081</th>\n",
       "      <td>AM1</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>1.547170</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187296</th>\n",
       "      <td>YF1</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>2.283271</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7746859</th>\n",
       "      <td>YF1</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>1.980990</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7790066</th>\n",
       "      <td>YF2</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7833273</th>\n",
       "      <td>YM1</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>3.091000</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7876480</th>\n",
       "      <td>YM2</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>0.906683</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7919687</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>0.254103</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample          gene_id       cpm           celltype\n",
       "2436       AF1  ENSG00000143452  0.556982  L4-5IT_RORB_TSHZ2\n",
       "48651      AF2  ENSG00000143452  1.850423  L4-5IT_RORB_TSHZ2\n",
       "94866      AF3  ENSG00000143452  0.565452  L4-5IT_RORB_TSHZ2\n",
       "141081     AM1  ENSG00000143452  1.547170  L4-5IT_RORB_TSHZ2\n",
       "187296     YF1  ENSG00000143452  2.283271  L4-5IT_RORB_TSHZ2\n",
       "...        ...              ...       ...                ...\n",
       "7746859    YF1  ENSG00000143452  1.980990   L56NP_TLE4_TSHZ2\n",
       "7790066    YF2  ENSG00000143452  0.000000   L56NP_TLE4_TSHZ2\n",
       "7833273    YM1  ENSG00000143452  3.091000   L56NP_TLE4_TSHZ2\n",
       "7876480    YM2  ENSG00000143452  0.906683   L56NP_TLE4_TSHZ2\n",
       "7919687    YM3  ENSG00000143452  0.254103   L56NP_TLE4_TSHZ2\n",
       "\n",
       "[180 rows x 4 columns]"
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_sig = sig_df.iloc[0]\n",
    "curr_gene_df = me_gene[me_gene['contrast'] == curr_sig['contrast']]\n",
    "curr_cpm_df = cpm[cpm['gene_id'] == curr_gene_df['gene_id'].values[0]]\n",
    "curr_cpm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "id": "05519299-7f5e-4ee0-9202-0e3eb1ffe3fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "      <th>contrast</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150698060</td>\n",
       "      <td>150720895</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>HORMAD1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>150718974</td>\n",
       "      <td>150719293</td>\n",
       "      <td>CZI_abs_64</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>absence</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>full_length</td>\n",
       "      <td>319</td>\n",
       "      <td>CZI_abs_64_ENSG00000143452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  gene_start   gene_end gene_strand          gene_id gene_name  \\\n",
       "7  chr1   150698060  150720895           -  ENSG00000143452   HORMAD1   \n",
       "\n",
       "        gene_type   me_start     me_end       me_id me_strand me_class het  \\\n",
       "7  protein_coding  150718974  150719293  CZI_abs_64         +      alu   .   \n",
       "\n",
       "  hom  me_type het_hom_ratio genotype insertion_category  length  \\\n",
       "7   .  absence             .        .        full_length     319   \n",
       "\n",
       "                     contrast  \n",
       "7  CZI_abs_64_ENSG00000143452  "
      ]
     },
     "execution_count": 362,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_gene[me_gene['contrast'] == curr_sig['contrast']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "id": "47a73daa-c102-4d61-a7f0-e642167aa1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "me_gene['contrast'] = [me_gene.iloc[i]['me_id'] + '_' + me_gene.iloc[i]['gene_id'] for i in range(len(me_gene))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b1156ad-6a03-42e2-8ccd-950db53b79e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def celltype_designmatrix():\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f753314-41ef-4c6e-b11a-083d63bd327a",
   "metadata": {},
   "outputs": [],
   "source": [
    "curr = pd.DataFrame(me_gene.iloc[i]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "488d40cb-5f5d-44b5-839e-9c68191a5682",
   "metadata": {},
   "outputs": [],
   "source": [
    "cpm_gene = cpm[cpm['gene_id'] == curr['gene_id'].values[0]]\n",
    "#skip genes that are not included in the filtered list\n",
    "if len(cpm_gene) > 0:\n",
    "    assert len(cpm_gene) == len(cpm_gene['celltype'].unique())*len(samples), 'some samples missing' #check that the correct n samples are there\n",
    "\n",
    "    #define design matrix for model\n",
    "    #============================\n",
    "    cpm_gene.set_index(['celltype', 'sample'], inplace=True)\n",
    "    cpm_gene.drop(columns='gene_id', inplace=True)\n",
    "    cpm_gene['age'] = [i[0] for i in cpm_gene.index.get_level_values('sample')]\n",
    "    cpm_gene['sex'] = [i[1] for i in cpm_gene.index.get_level_values('sample')]\n",
    "    cpm_gene['age_encoded'] = cpm_gene['age'].map({'Y': 0, 'A': 1})\n",
    "    cpm_gene['sex_encoded'] = cpm_gene['sex'].map({'M': 0, 'F': 1})\n",
    "    \n",
    "    #define me_type values, 0,1,2\n",
    "    if curr['me_type'].values == 'insertion': \n",
    "        hom_val = 2\n",
    "        wt_val = 0\n",
    "    if curr['me_type'].values == 'absence':\n",
    "        hom_val = 0\n",
    "        wt_val = 2\n",
    "    vec = np.full(len(cpm_gene), np.nan)\n",
    "    vec[cpm_gene.index.get_level_values('sample').str.contains(curr['het'].values[0].replace(',','|'))] = 1\n",
    "    vec[cpm_gene.index.get_level_values('sample').str.contains(curr['hom'].values[0].replace(',','|'))] = hom_val\n",
    "    wt = ','.join(samples[np.in1d(samples, np.append(curr['het'].values[0].split(','), \n",
    "                                                    curr['hom'].values[0].split(','))) == False]).replace(',','|')\n",
    "    if len(wt) > 0: vec[cpm_gene.index.get_level_values('sample').str.contains(wt)] = wt_val\n",
    "    cpm_gene['me_type'] = vec\n",
    "\n",
    "    #run model\n",
    "    #==========\n",
    "    out = cpm_gene.groupby('celltype').apply(pfn.run_regression).reset_index()\n",
    "    out['contrast'] = f\"{curr['me_id'].values[0]}_{curr['gene_id'].values[0]}\"\n",
    "    out_df = pd.concat([out_df, out])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "26891568-593f-480e-b342-42fd87a7ce98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cpm</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_encoded</th>\n",
       "      <th>sex_encoded</th>\n",
       "      <th>me_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L4-5IT_RORB_TSHZ2</th>\n",
       "      <th>AF1</th>\n",
       "      <td>0.383286</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF2</th>\n",
       "      <td>1.516757</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF3</th>\n",
       "      <td>0.376111</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM1</th>\n",
       "      <td>0.783174</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF1</th>\n",
       "      <td>0.273898</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L56NP_TLE4_TSHZ2</th>\n",
       "      <th>YF1</th>\n",
       "      <td>0.378245</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF2</th>\n",
       "      <td>0.427399</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM1</th>\n",
       "      <td>0.481128</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM2</th>\n",
       "      <td>1.584667</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM3</th>\n",
       "      <td>0.174273</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cpm age sex  age_encoded  sex_encoded  me_type\n",
       "celltype          sample                                                     \n",
       "L4-5IT_RORB_TSHZ2 AF1     0.383286   A   F            1            1      0.0\n",
       "                  AF2     1.516757   A   F            1            1      0.0\n",
       "                  AF3     0.376111   A   F            1            1      0.0\n",
       "                  AM1     0.783174   A   M            1            0      0.0\n",
       "                  YF1     0.273898   Y   F            0            1      0.0\n",
       "...                            ...  ..  ..          ...          ...      ...\n",
       "L56NP_TLE4_TSHZ2  YF1     0.378245   Y   F            0            1      0.0\n",
       "                  YF2     0.427399   Y   F            0            1      0.0\n",
       "                  YM1     0.481128   Y   M            0            0      0.0\n",
       "                  YM2     1.584667   Y   M            0            0      0.0\n",
       "                  YM3     0.174273   Y   M            0            0      0.0\n",
       "\n",
       "[180 rows x 6 columns]"
      ]
     },
     "execution_count": 352,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#by celltype, show each significant gene\n",
    "\n",
    "#group genes together for each celltype (big up/down), show joint plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3dfc47-4846-4313-b8f7-4033c380d48c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#by gene, show all cells, identify celltype differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "544c2df0-9a66-4044-8130-e4104da5cc44",
   "metadata": {},
   "outputs": [],
   "source": [
    "#do stats of insertion type, coefficients by insertion location -> i.e. where in the gene; by full length; by l1 vs alu\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "e51e37f6-7314-442c-a2b2-f2bc63cc4df9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CZI_abs_64_ENSG00000143452'"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_df['contrast'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "7c19cc35-a8c4-40a5-bba9-e634903d12c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>level_1</th>\n",
       "      <th>coefficients</th>\n",
       "      <th>p_values</th>\n",
       "      <th>contrast</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGE_ADARB2_ADAM33</td>\n",
       "      <td>const</td>\n",
       "      <td>0.177444</td>\n",
       "      <td>0.772504</td>\n",
       "      <td>CZI_abs_7_ENSG00000230703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGE_ADARB2_ADAM33</td>\n",
       "      <td>age_encoded</td>\n",
       "      <td>0.172562</td>\n",
       "      <td>0.658607</td>\n",
       "      <td>CZI_abs_7_ENSG00000230703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGE_ADARB2_ADAM33</td>\n",
       "      <td>sex_encoded</td>\n",
       "      <td>0.172562</td>\n",
       "      <td>0.658607</td>\n",
       "      <td>CZI_abs_7_ENSG00000230703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGE_ADARB2_ADAM33</td>\n",
       "      <td>me_type</td>\n",
       "      <td>-0.099603</td>\n",
       "      <td>0.869062</td>\n",
       "      <td>CZI_abs_7_ENSG00000230703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGE_LAMP5</td>\n",
       "      <td>const</td>\n",
       "      <td>0.413008</td>\n",
       "      <td>0.034756</td>\n",
       "      <td>CZI_abs_7_ENSG00000230703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>MGE_SST</td>\n",
       "      <td>me_type</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CZI_abs_1499_ENSG00000230676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>MGE_SST_CLMP</td>\n",
       "      <td>const</td>\n",
       "      <td>0.783821</td>\n",
       "      <td>0.011151</td>\n",
       "      <td>CZI_abs_1499_ENSG00000230676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>MGE_SST_CLMP</td>\n",
       "      <td>age_encoded</td>\n",
       "      <td>-0.247380</td>\n",
       "      <td>0.429813</td>\n",
       "      <td>CZI_abs_1499_ENSG00000230676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MGE_SST_CLMP</td>\n",
       "      <td>sex_encoded</td>\n",
       "      <td>-0.493100</td>\n",
       "      <td>0.142570</td>\n",
       "      <td>CZI_abs_1499_ENSG00000230676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MGE_SST_CLMP</td>\n",
       "      <td>me_type</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CZI_abs_1499_ENSG00000230676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9584 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             celltype      level_1  coefficients  p_values  \\\n",
       "0   CGE_ADARB2_ADAM33        const      0.177444  0.772504   \n",
       "1   CGE_ADARB2_ADAM33  age_encoded      0.172562  0.658607   \n",
       "2   CGE_ADARB2_ADAM33  sex_encoded      0.172562  0.658607   \n",
       "3   CGE_ADARB2_ADAM33      me_type     -0.099603  0.869062   \n",
       "4           CGE_LAMP5        const      0.413008  0.034756   \n",
       "..                ...          ...           ...       ...   \n",
       "75            MGE_SST      me_type      0.000000       NaN   \n",
       "76       MGE_SST_CLMP        const      0.783821  0.011151   \n",
       "77       MGE_SST_CLMP  age_encoded     -0.247380  0.429813   \n",
       "78       MGE_SST_CLMP  sex_encoded     -0.493100  0.142570   \n",
       "79       MGE_SST_CLMP      me_type      0.000000       NaN   \n",
       "\n",
       "                        contrast  \n",
       "0      CZI_abs_7_ENSG00000230703  \n",
       "1      CZI_abs_7_ENSG00000230703  \n",
       "2      CZI_abs_7_ENSG00000230703  \n",
       "3      CZI_abs_7_ENSG00000230703  \n",
       "4      CZI_abs_7_ENSG00000230703  \n",
       "..                           ...  \n",
       "75  CZI_abs_1499_ENSG00000230676  \n",
       "76  CZI_abs_1499_ENSG00000230676  \n",
       "77  CZI_abs_1499_ENSG00000230676  \n",
       "78  CZI_abs_1499_ENSG00000230676  \n",
       "79  CZI_abs_1499_ENSG00000230676  \n",
       "\n",
       "[9584 rows x 5 columns]"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "f2ce353d-55c1-4c5b-9610-21a5c940ecc8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24066774</td>\n",
       "      <td>24083565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230703</td>\n",
       "      <td>MYOM3-AS1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>24076520</td>\n",
       "      <td>24076812</td>\n",
       "      <td>CZI_abs_7</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>AM3,YF2</td>\n",
       "      <td>absence</td>\n",
       "      <td>4.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55915603</td>\n",
       "      <td>55944980</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230250</td>\n",
       "      <td>LINC01753</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>55935619</td>\n",
       "      <td>55935930</td>\n",
       "      <td>CZI_abs_19</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56154545</td>\n",
       "      <td>56477687</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260971</td>\n",
       "      <td>ENSG00000260971</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56173433</td>\n",
       "      <td>56524495</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284686</td>\n",
       "      <td>ENSG00000284686</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57605847</td>\n",
       "      <td>57606206</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000236888</td>\n",
       "      <td>RPS20P5</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>57606184</td>\n",
       "      <td>57606197</td>\n",
       "      <td>CZI_ins_818</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AM3,YF1</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,YM1,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>0.2857142857142857</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>chr8</td>\n",
       "      <td>143636019</td>\n",
       "      <td>143656418</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000183309</td>\n",
       "      <td>ZNF623</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143641445</td>\n",
       "      <td>143641764</td>\n",
       "      <td>CZI_abs_1430</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF2,AM3,YM1,YM2</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>chr8</td>\n",
       "      <td>144772224</td>\n",
       "      <td>144787348</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000196378</td>\n",
       "      <td>ZNF34</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>144776325</td>\n",
       "      <td>144776913</td>\n",
       "      <td>CZI_abs_1433</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>9.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>chr9</td>\n",
       "      <td>35791003</td>\n",
       "      <td>35809732</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159899</td>\n",
       "      <td>NPR2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>35803110</td>\n",
       "      <td>35803402</td>\n",
       "      <td>CZI_abs_1464</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AM1,AM2,YF1,YF2,YM2,YM3</td>\n",
       "      <td>AF2,AF3,AM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>2.333333333333333</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>chr9</td>\n",
       "      <td>73042745</td>\n",
       "      <td>73056528</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000228024</td>\n",
       "      <td>CYP1D1P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>73054613</td>\n",
       "      <td>73054880</td>\n",
       "      <td>CZI_abs_1473</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129430652</td>\n",
       "      <td>129451422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230676</td>\n",
       "      <td>ENSG00000230676</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129435084</td>\n",
       "      <td>129435407</td>\n",
       "      <td>CZI_abs_1499</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr  gene_start   gene_end gene_strand          gene_id  \\\n",
       "0    chr1    24066774   24083565           +  ENSG00000230703   \n",
       "1    chr1    55915603   55944980           -  ENSG00000230250   \n",
       "2    chr1    56154545   56477687           -  ENSG00000260971   \n",
       "3    chr1    56173433   56524495           -  ENSG00000284686   \n",
       "4    chr1    57605847   57606206           -  ENSG00000236888   \n",
       "..    ...         ...        ...         ...              ...   \n",
       "122  chr8   143636019  143656418           +  ENSG00000183309   \n",
       "123  chr8   144772224  144787348           -  ENSG00000196378   \n",
       "124  chr9    35791003   35809732           +  ENSG00000159899   \n",
       "125  chr9    73042745   73056528           +  ENSG00000228024   \n",
       "126  chr9   129430652  129451422           -  ENSG00000230676   \n",
       "\n",
       "           gene_name               gene_type   me_start     me_end  \\\n",
       "0          MYOM3-AS1                  lncRNA   24076520   24076812   \n",
       "1          LINC01753                  lncRNA   55935619   55935930   \n",
       "2    ENSG00000260971                  lncRNA   56365451   56369289   \n",
       "3    ENSG00000284686          protein_coding   56365451   56369289   \n",
       "4            RPS20P5    processed_pseudogene   57606184   57606197   \n",
       "..               ...                     ...        ...        ...   \n",
       "122           ZNF623          protein_coding  143641445  143641764   \n",
       "123            ZNF34          protein_coding  144776325  144776913   \n",
       "124             NPR2          protein_coding   35803110   35803402   \n",
       "125          CYP1D1P  unprocessed_pseudogene   73054613   73054880   \n",
       "126  ENSG00000230676                  lncRNA  129435084  129435407   \n",
       "\n",
       "            me_id me_strand me_class                                  het  \\\n",
       "0       CZI_abs_7         -      alu      AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3   \n",
       "1      CZI_abs_19         +      alu                                    .   \n",
       "2    CZI_abs_1543         +       l1                                  AF3   \n",
       "3    CZI_abs_1543         +       l1                                  AF3   \n",
       "4     CZI_ins_818         -      alu                              AM3,YF1   \n",
       "..            ...       ...      ...                                  ...   \n",
       "122  CZI_abs_1430         -      alu                                    .   \n",
       "123  CZI_abs_1433         +      alu  AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3   \n",
       "124  CZI_abs_1464         -      alu          AF1,AM1,AM2,YF1,YF2,YM2,YM3   \n",
       "125  CZI_abs_1473         +      alu                                    .   \n",
       "126  CZI_abs_1499         -      alu                                    .   \n",
       "\n",
       "                                             hom    me_type  \\\n",
       "0                                        AM3,YF2    absence   \n",
       "1                                            AF1    absence   \n",
       "2                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "3                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "4                    AF1,AF2,AF3,AM1,AM2,YM1,YM3  insertion   \n",
       "..                                           ...        ...   \n",
       "122                              AF2,AM3,YM1,YM2    absence   \n",
       "123                                          YM1    absence   \n",
       "124                                  AF2,AF3,AM3    absence   \n",
       "125                                          YM1    absence   \n",
       "126  AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3    absence   \n",
       "\n",
       "          het_hom_ratio genotype insertion_category  length  \n",
       "0                   4.0      het        full_length     292  \n",
       "1                  -inf      hom        full_length     311  \n",
       "2                   0.2      hom          truncated    3838  \n",
       "3                   0.2      hom          truncated    3838  \n",
       "4    0.2857142857142857      hom          truncated      13  \n",
       "..                  ...      ...                ...     ...  \n",
       "122                -inf      hom        full_length     319  \n",
       "123                 9.0      het        full_length     588  \n",
       "124   2.333333333333333      het        full_length     292  \n",
       "125                -inf      hom          truncated     267  \n",
       "126                -inf      hom        full_length     323  \n",
       "\n",
       "[127 rows x 19 columns]"
      ]
     },
     "execution_count": 349,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "3568f818-2769-4a94-9bf4-658e1d9ca2e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cpm</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>0.729183</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>0.069128</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000240361</td>\n",
       "      <td>0.147933</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF1</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>0.055565</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7960576</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>8.790963</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7960577</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>2.029100</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7960578</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>10.021181</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7960579</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>0.535441</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7960580</th>\n",
       "      <td>YM3</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>1.191537</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7960581 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample          gene_id        cpm           celltype\n",
       "0          AF1  ENSG00000223972   0.000000  L4-5IT_RORB_TSHZ2\n",
       "1          AF1  ENSG00000227232   0.729183  L4-5IT_RORB_TSHZ2\n",
       "2          AF1  ENSG00000243485   0.069128  L4-5IT_RORB_TSHZ2\n",
       "3          AF1  ENSG00000240361   0.147933  L4-5IT_RORB_TSHZ2\n",
       "4          AF1  ENSG00000186092   0.055565  L4-5IT_RORB_TSHZ2\n",
       "...        ...              ...        ...                ...\n",
       "7960576    YM3  ENSG00000198695   8.790963   L56NP_TLE4_TSHZ2\n",
       "7960577    YM3  ENSG00000210194   2.029100   L56NP_TLE4_TSHZ2\n",
       "7960578    YM3  ENSG00000198727  10.021181   L56NP_TLE4_TSHZ2\n",
       "7960579    YM3  ENSG00000210195   0.535441   L56NP_TLE4_TSHZ2\n",
       "7960580    YM3  ENSG00000210196   1.191537   L56NP_TLE4_TSHZ2\n",
       "\n",
       "[7960581 rows x 4 columns]"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f2e3bd-6044-4d26-b3ad-79320ba3686d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb67e772-6555-4592-bec4-a7b26aee1f3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset genes -> define distance, output column with list of genes to test -> just get csv of each gene -> me interaction\n",
    "\n",
    "\n",
    "#define model, based on het/hom and absence/insertion info; get full sample list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27a2c7fc-3bce-40fb-b959-c1949b0c0a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualise significant effects, by celltype, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2938ba-9163-4cd1-a42c-7c2b2343e46c",
   "metadata": {},
   "source": [
    "# metheQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ef7a6b-16ef-432d-ac80-cd038a611adf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
