{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cd163de-4970-4633-b1a8-9ab89a24cf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from glob import glob\n",
    "from scipy.stats import permutation_test, false_discovery_control, spearmanr\n",
    "import pybedtools\n",
    "import gene_expression\n",
    "import process as pfn\n",
    "\n",
    "from warnings import filterwarnings\n",
    "filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "filterwarnings(\"ignore\", category=pd.errors.SettingWithCopyWarning)\n",
    "%load_ext autoreload\n",
    "#set up paths\n",
    "#=========================\n",
    "workdir = '/home/AD/rkgadde/L1IP'\n",
    "annotation = f'{workdir}/hg38_data/annotations/gencode.v46.basic.annotation.sorted.genes.gff3'\n",
    "celltypes = f'{workdir}/celltypes.txt'\n",
    "\n",
    "genedir = f'{workdir}/gene_data'\n",
    "figdir = f'{workdir}/results/CZI/plots'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9163b55f-80a4-4dc2-9a76-8e4eb311fbdf",
   "metadata": {},
   "source": [
    "# Load in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8079bced-2a7c-49aa-8ca9-d81b160fb29e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in mes\n",
    "comb_df = pd.read_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', index_col=0)\n",
    "samples = pd.read_csv(f'{workdir}/samples.txt', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bcb723b-cd43-464a-a57f-27d383dc2482",
   "metadata": {},
   "outputs": [],
   "source": [
    "#process mes into joint bed file\n",
    "#process l1\n",
    "me_type = 'L1'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "l1_df = pd.concat([abs_df, ins_df])\n",
    "l1_df['class'] = 'l1'\n",
    "\n",
    "#process alu\n",
    "me_type = 'Alu'\n",
    "abs_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_abs.tsv'\n",
    "ins_file = f'{workdir}/mC_data/CZI/type/vars/all_{me_type}_ins.tsv'\n",
    "abs_df = pd.read_csv(abs_file, sep='\\t', usecols=[0,1,2,3,4,5,6,7,8], names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "abs_df['me_type'] = 'absence'\n",
    "ins_df = pd.read_csv(ins_file, sep='\\t', usecols=[0,1,2,3,4,5,6, 7,8],  names=['chrom','start','end','id', 'length', 'strand', 'class', 'het', 'hom'])[1:]\n",
    "ins_df['me_type'] = 'insertion'\n",
    "alu_df = pd.concat([abs_df, ins_df])\n",
    "alu_df['class'] = 'alu'\n",
    "\n",
    "#add labels\n",
    "comb_df = pd.concat([l1_df, alu_df])\n",
    "comb_df['start'] = comb_df['start'].astype(int)\n",
    "comb_df['end'] = comb_df['end'].astype(int)\n",
    "comb_df['length'] = comb_df['end'] - comb_df['start']\n",
    "comb_df['het'] = comb_df['het'].fillna('NaN').astype(str)\n",
    "comb_df['hom'] = comb_df['hom'].fillna('NaN').astype(str)\n",
    "comb_df.reset_index(drop=True, inplace=True)\n",
    "het_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['het']])\n",
    "hom_bool = np.asarray([True if 'NaN' not in i else False for i in comb_df['hom']])\n",
    "het_count = np.asarray([len(i.split(',')) for i in comb_df['het']])\n",
    "hom_count = np.asarray([len(i.split(',')) for i in comb_df['hom']])\n",
    "ratio = het_count/hom_count\n",
    "#set to NaN and inf\n",
    "ratio[(het_bool == False) & (hom_bool==False)] = np.nan\n",
    "ratio[(het_bool == True) & (hom_bool==False)] = np.inf\n",
    "ratio[(het_bool == False) & (hom_bool==True)] = -1*np.inf\n",
    "comb_df['het_over_hom'] = ratio\n",
    "\n",
    "# label as majority het or hom, if >2x\n",
    "genotype = np.empty(len(comb_df)).astype(str)\n",
    "genotype[ratio >= 2] = 'het'\n",
    "genotype[ratio <= 0.5] = 'hom'\n",
    "genotype[(ratio <= 2) & (ratio >= 0.5)] = 'mixed'\n",
    "genotype[np.isnan(ratio)] = 'NaN'\n",
    "comb_df['genotype'] = genotype\n",
    "#label as truncated or full length\n",
    "trunc = np.empty(len(comb_df), dtype=object)\n",
    "l1_mask = comb_df['class'] == 'l1'\n",
    "trunc[l1_mask & (comb_df['length'] > 5500)] = 'full_length'\n",
    "trunc[l1_mask & (comb_df['length'] <= 5500)] = 'truncated'\n",
    "alu_mask = comb_df['class'] == 'alu'\n",
    "trunc[alu_mask & (comb_df['length'] > 280)] = 'full_length'\n",
    "trunc[alu_mask & (comb_df['length'] <= 280)] = 'truncated'\n",
    "comb_df['insertion_category'] = trunc\n",
    "\n",
    "comb_df.to_csv('/cndd3/dburrows/DATA/me_polymorphisms/analysis/comb_polymorphic_me.csv', sep='\\t', header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef9501d-61f3-4a14-a826-84eb828a2645",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#Get bed file with gene names\n",
    "#==================================\n",
    "cd /cndd3/dburrows/DATA/annotations/gencode/\n",
    "#Split up 4th column into gene id and type\n",
    "# awk 'BEGIN{OFS=\"\\t\"} {split($4, a, \"_\"); $4 = a[1]; $5 = a[2]; print}' gencode.v44.annotation.hg38.genelabels.bed > tmp_split.bed\n",
    "# mv tmp_split.bed gencode.v44.annotation.hg38.genelabels.bed\n",
    "\n",
    "#remove duplicated gene loci (exons from isoforms) that have the same chromosome, start+end position gene_label, and strand\n",
    "# awk '!seen[$1,$2,$3,$5,$6]++' gencode.v44.annotation.hg38.genelabels.bed > tmp_seen.bed\n",
    "# mv tmp_seen.bed gencode.v44.annotation.hg38.genelabels.unique.bed\n",
    "awk -F'\\t' '{\n",
    "    # Extract necessary fields from the 9th column (attributes)\n",
    "    match($9, /gene_id=([^;]+)/, gene_id);\n",
    "    match($9, /gene_name=([^;]+)/, gene_name);\n",
    "    match($9, /gene_type=([^;]+)/, gene_type);\n",
    "\n",
    "    # Rearrange and print the columns\n",
    "    print $1, $4, $5, $2, $3, $7, gene_id[1], gene_name[1], gene_type[1];\n",
    "}' OFS=\"\\t\" gencode.v46.basic.annotation.sorted.genes.gff3 > gencode.v46.basic.annotation.sorted.genes.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc721cc9-973a-4396-9c54-9d6fd2a7c0a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection of mes with genes\n",
    "#bedconvert and load annotations\n",
    "me_bt = pybedtools.BedTool.from_dataframe(comb_df)\n",
    "gene_df = pd.read_csv('/cndd3/dburrows/DATA/annotations/gencode/gencode.v46.basic.annotation.sorted.genes.bed', sep='\\t',header=None)\n",
    "gene_bt = pybedtools.BedTool.from_dataframe(gene_df)\n",
    "\n",
    "#compute intersection\n",
    "me_gene = gene_bt.intersect(me_bt, f=0.01, wo=True) \n",
    "me_gene = me_gene.to_dataframe(disable_auto_names=True, header=None)\n",
    "me_gene.rename(columns={0: 'chr', 1: 'gene_start', 2: 'gene_end', 5: 'gene_strand',\n",
    "                        6: 'gene_id', 7: 'gene_name', 8: 'gene_type', 10: 'me_start',\n",
    "                       11: 'me_end', 12: 'me_id', 14: 'me_strand', 15: 'me_class', 16: 'het',\n",
    "                       17: 'hom', 18: 'me_type', 19: 'het_hom_ratio', 20:'genotype', \n",
    "                       21: 'insertion_category', 22: 'length'}, inplace=True)\n",
    "me_gene.drop(columns=np.asarray(me_gene.columns[(np.asarray([type(i) for i in me_gene.columns]) == int)]), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1918d864-9099-4863-863f-c921e9419b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find intersection with genomic regions\n",
    "\n",
    "#compute intersection\n",
    "l1_cgi = l1_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_cgi = alu_bt.intersect(cgi_bt, f=0.3, wo=True) \n",
    "alu_all = alu_bt.intersect(all_bt, f=0.3, wo=True) \n",
    "l1_all = all_bt.intersect(l1_bt, f=0.3, wo=True) \n",
    "\n",
    "l1_cgi_df = l1_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_cgi_df = alu_cgi.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_cgi_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "alu_all_df = alu_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "alu_all_df.rename(columns={16: 'region', 9: 'me_type', 11: 'genotype', 12: 'insertion_category', 7: 'het', 8: 'hom', 4: 'length'}, inplace=True)\n",
    "l1_all_df = l1_all.to_dataframe(disable_auto_names=True, header=None)\n",
    "l1_all_df.rename(columns={3: 'region', 15: 'me_type', 17: 'genotype', 18: 'insertion_category', 13: 'het', 14: 'hom', 10: 'length'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3735c321-84a2-4c18-a945-e1ca569e2b23",
   "metadata": {},
   "source": [
    "# eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb6af47-e48c-4fcb-b2da-a92924dce8ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run for MEs directly overlapping genes\n",
    "#=====================================\n",
    "\n",
    "#load in + filter cpm data\n",
    "me_gene['gene_id'] = np.asarray([i.split('.')[0] for i in me_gene['gene_id']])\n",
    "cpm_path = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "cpm = pfn.combine_genes_bycell_filter(rna = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , mini = 0\n",
    "                                      , n_samp = 0)\n",
    "\n",
    "samples = cpm['sample'].unique()\n",
    "\n",
    "# run model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d5b36a00-3dd5-4a02-b6bb-9100f1770fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bb47f61f-0837-4388-9e62-1799656df0fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<statsmodels.regression.linear_model.RegressionResultsWrapper at 0x7ff2dea0feb0>"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import statsmodels.api as sm\n",
    "U = np.random.rand(100)  # Gene expression\n",
    "Y = np.random.choice([0, 1, 2], 100)  # Genotype (none, het, hom)\n",
    "B = np.random.rand(100, 3)  # 3 confounders/predictors\n",
    "\n",
    "# Add genotype and intercept to predictors\n",
    "X = np.column_stack([Y, B])\n",
    "X = sm.add_constant(X)\n",
    "\n",
    "# Fit the model\n",
    "model = sm.OLS(U, X).fit()\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "2170c2ed-6fc8-4687-910d-e96f728b465c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run model\n",
    "#==========\n",
    "\n",
    "# extract each gene and celltype\n",
    "i = 0\n",
    "curr = pd.DataFrame(me_gene.iloc[0]).T\n",
    "cpm_unq = cpm['gene_id'].unique()\n",
    "cpm_gene = cpm[cpm['gene_id'] == curr['gene_id'].values[0]]\n",
    "if len(cpm_gene) > 0:\n",
    "    assert len(cpm_gene) == len(cpm_gene['celltype'].unique())*len(samples), 'some samples missing' #check that the correct n samples are there\n",
    "    celltypes = cpm_gene['celltype'].unique()\n",
    "    for cell in celltypes:\n",
    "        1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b08d0a8a-261d-4a75-af6a-d902fa8f78bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y',\n",
       " 'Y']"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i[0] for i in cpm_gene.index.get_level_values('sample')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a0b88bd3-ea65-4ff3-be07-70f98723bd99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cpm</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L4-5IT_RORB_TSHZ2</th>\n",
       "      <th>AF1</th>\n",
       "      <td>0.028050</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF2</th>\n",
       "      <td>0.376107</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM1</th>\n",
       "      <td>0.469351</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF1</th>\n",
       "      <td>0.063228</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">L56NP_TLE4_TSHZ2</th>\n",
       "      <th>YF1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YF2</th>\n",
       "      <td>0.119292</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM1</th>\n",
       "      <td>0.337871</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YM3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>Y</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cpm age sex\n",
       "celltype          sample                  \n",
       "L4-5IT_RORB_TSHZ2 AF1     0.028050   A   F\n",
       "                  AF2     0.376107   A   F\n",
       "                  AF3     0.000000   A   F\n",
       "                  AM1     0.469351   A   M\n",
       "                  YF1     0.063228   Y   F\n",
       "...                            ...  ..  ..\n",
       "L56NP_TLE4_TSHZ2  YF1     0.000000   Y   F\n",
       "                  YF2     0.119292   Y   F\n",
       "                  YM1     0.337871   Y   M\n",
       "                  YM2     0.000000   Y   M\n",
       "                  YM3     0.000000   Y   M\n",
       "\n",
       "[180 rows x 3 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cpm_gene.set_index(['celltype', 'sample'], inplace=True)\n",
    "# cpm_gene.drop(columns='gene_id', inplace=True)\n",
    "cpm_gene['age'] = [i[0] for i in cpm_gene.index.get_level_values('sample')]\n",
    "cpm_gene['sex'] = [i[1] for i in cpm_gene.index.get_level_values('sample')]\n",
    "cpm_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58857da4-794f-4f02-9fee-308fc3a5d4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import statsmodels.api as sm\n",
    "U =  # Gene expression\n",
    "Y = np.random.choice([0, 1, 2], 100)  # Genotype (none, het, hom)\n",
    "B = np.random.rand(100, 3)  # 3 confounders/predictors\n",
    "\n",
    "# Add genotype and intercept to predictors\n",
    "X = np.column_stack([Y, B])\n",
    "X = sm.add_constant(X)\n",
    "\n",
    "# Fit the model\n",
    "model = sm.OLS(U, X).fit()\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "e12e9509-d0e2-45bf-b844-ef600bdc323f",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of ['celltype', 'sample'] are in the columns\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1744016/2366847994.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcpm_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'celltype'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'sample'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mcpm_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'gene_id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mcpm_gene\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/mambaforge/lib/python3.10/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, keys, drop, append, inplace, verify_integrity)\u001b[0m\n\u001b[1;32m   6105\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mfound\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6106\u001b[0m                         \u001b[0mmissing\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6107\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6108\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmissing\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6109\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"None of {missing} are in the columns\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6110\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6111\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6112\u001b[0m             \u001b[0mframe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of ['celltype', 'sample'] are in the columns\""
     ]
    }
   ],
   "source": [
    "cpm_gene.set_index(['celltype', 'sample'], inplace=True)\n",
    "cpm_gene.drop(columns='gene_id', inplace=True)\n",
    "cpm_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2dd323e-4e82-46f6-ac71-a501c0c5627a",
   "metadata": {},
   "outputs": [],
   "source": [
    "U = #gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "53ac2fc3-38a0-48be-a468-0d81083b2894",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24066774</td>\n",
       "      <td>24083565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230703</td>\n",
       "      <td>MYOM3-AS1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>24076520</td>\n",
       "      <td>24076812</td>\n",
       "      <td>CZI_abs_7</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>AM3,YF2</td>\n",
       "      <td>absence</td>\n",
       "      <td>4.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr gene_start  gene_end gene_strand          gene_id  gene_name  \\\n",
       "0  chr1   24066774  24083565           +  ENSG00000230703  MYOM3-AS1   \n",
       "\n",
       "  gene_type  me_start    me_end      me_id me_strand me_class  \\\n",
       "0    lncRNA  24076520  24076812  CZI_abs_7         -      alu   \n",
       "\n",
       "                               het      hom  me_type het_hom_ratio genotype  \\\n",
       "0  AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3  AM3,YF2  absence           4.0      het   \n",
       "\n",
       "  insertion_category length  \n",
       "0        full_length    292  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ecc0eb66-fe08-4104-ae40-c8d84b3f4b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run for MEs directly overlapping genes\n",
    "#=====================================\n",
    "\n",
    "#load in + filter cpm data\n",
    "me_gene['gene_id'] = np.asarray([i.split('.')[0] for i in me_gene['gene_id']])\n",
    "cpm_path = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "cpm = pfn.combine_genes_bycell_filter(rna = '/cndd2/jchien/project/CZI_human/pseudobulk_rna/logcpm/*_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , pattern = '_combinesample_log2cpm+1.tsv.gz'\n",
    "                                      , mini = 0\n",
    "                                      , n_samp = 0)\n",
    "\n",
    "samples = cpm['sample'].unique()\n",
    "\n",
    "# run model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "21bb7c9e-ceef-4853-8067-d7ad663ae1d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr15</td>\n",
       "      <td>70654554</td>\n",
       "      <td>70763558</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000137831</td>\n",
       "      <td>UACA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>70729509</td>\n",
       "      <td>70735169</td>\n",
       "      <td>CZI_abs_1585</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3</td>\n",
       "      <td>YF1,YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>4.5</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>5660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr gene_start  gene_end gene_strand          gene_id gene_name  \\\n",
       "38  chr15   70654554  70763558           -  ENSG00000137831      UACA   \n",
       "\n",
       "         gene_type  me_start    me_end         me_id me_strand me_class  \\\n",
       "38  protein_coding  70729509  70735169  CZI_abs_1585         +       l1   \n",
       "\n",
       "                                    het      hom  me_type het_hom_ratio  \\\n",
       "38  AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3  YF1,YM1  absence           4.5   \n",
       "\n",
       "   genotype insertion_category length  \n",
       "38      het        full_length   5660  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prac = pd.DataFrame(me_gene.iloc[i]).T\n",
    "prac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "5a477070-2511-4cf8-9021-decf501fcd47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       ENSG00000230703.1\n",
       "1       ENSG00000230250.3\n",
       "2       ENSG00000260971.4\n",
       "3       ENSG00000284686.1\n",
       "4       ENSG00000236888.1\n",
       "              ...        \n",
       "122    ENSG00000183309.12\n",
       "123    ENSG00000196378.14\n",
       "124    ENSG00000159899.16\n",
       "125     ENSG00000228024.1\n",
       "126     ENSG00000230676.1\n",
       "Name: gene_id, Length: 127, dtype: object"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_gene['gene_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "a6df2075-f58c-4ad1-bfe1-ea9fae69a477",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          ENSG00000238009\n",
       "1          ENSG00000228463\n",
       "2          ENSG00000237094\n",
       "3          ENSG00000230021\n",
       "4          ENSG00000225972\n",
       "                ...       \n",
       "3975718    ENSG00000198695\n",
       "3975719    ENSG00000210194\n",
       "3975720    ENSG00000198727\n",
       "3975721    ENSG00000210195\n",
       "3975722    ENSG00000210196\n",
       "Name: gene_id, Length: 3975723, dtype: object"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm['gene_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "4ddcd6d3-f463-4b99-9442-4e901ef2446e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">AF1</th>\n",
       "      <th>ENSG00000238009.6</th>\n",
       "      <td>2.211659</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463.10</th>\n",
       "      <td>4.014924</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237094.12</th>\n",
       "      <td>1.947321</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000230021.10</th>\n",
       "      <td>6.679229</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000225972.1</th>\n",
       "      <td>0.676991</td>\n",
       "      <td>L4-5IT_RORB_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">YM3</th>\n",
       "      <th>ENSG00000198695.2</th>\n",
       "      <td>8.790963</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210194.1</th>\n",
       "      <td>2.029100</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727.2</th>\n",
       "      <td>10.021181</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210195.2</th>\n",
       "      <td>0.535441</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210196.2</th>\n",
       "      <td>1.191537</td>\n",
       "      <td>L56NP_TLE4_TSHZ2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3975723 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                0           celltype\n",
       "AF1 ENSG00000238009.6    2.211659  L4-5IT_RORB_TSHZ2\n",
       "    ENSG00000228463.10   4.014924  L4-5IT_RORB_TSHZ2\n",
       "    ENSG00000237094.12   1.947321  L4-5IT_RORB_TSHZ2\n",
       "    ENSG00000230021.10   6.679229  L4-5IT_RORB_TSHZ2\n",
       "    ENSG00000225972.1    0.676991  L4-5IT_RORB_TSHZ2\n",
       "...                           ...                ...\n",
       "YM3 ENSG00000198695.2    8.790963   L56NP_TLE4_TSHZ2\n",
       "    ENSG00000210194.1    2.029100   L56NP_TLE4_TSHZ2\n",
       "    ENSG00000198727.2   10.021181   L56NP_TLE4_TSHZ2\n",
       "    ENSG00000210195.2    0.535441   L56NP_TLE4_TSHZ2\n",
       "    ENSG00000210196.2    1.191537   L56NP_TLE4_TSHZ2\n",
       "\n",
       "[3975723 rows x 2 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat(cpm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "ec951e35-38f6-448d-aa47-9fb4ea1618a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "rna = glob.glob(f'{rna_path}')\n",
    "\n",
    "matrices = []\n",
    "mini = np.log2(mini+1) #minimum log2cpm+1 value\n",
    "\n",
    "for rna_file in rna:\n",
    "    celltype = os.path.basename(rna_file).replace(pattern, '')\n",
    "    m = pd.read_csv(rna_file, sep='\\t', index_col=0)\n",
    "    m = m[np.sum(m > mini, axis=1) > n_samp]\n",
    "    m = pd.DataFrame(m.T.stack())\n",
    "    m['celltype'] = celltype\n",
    "    matrices.append(m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ae011e79-fc39-4013-8275-637267047567",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "171"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(cpm['gene_id'].isin([cpm['gene_id'].values[0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ed91042a-1842-4031-adea-ddb24adb3905",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000228463'"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm['gene_id'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e998f3ee-067d-4659-b83d-11f31d9710e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb67e772-6555-4592-bec4-a7b26aee1f3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset genes -> define distance, output column with list of genes to test -> just get csv of each gene -> me interaction\n",
    "\n",
    "\n",
    "#define model, based on het/hom and absence/insertion info; get full sample list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4e5d72-8b07-4aa3-a461-e632b3c7adf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#model:\n",
    "# U ~ XY + AiBi ...\n",
    "# U = gene expression for gene U\n",
    "# X = coefficient for locus\n",
    "# Y = 0,1 or 2 for none, het or hom\n",
    "# Ai = coefficient for predictor variable i\n",
    "# Bi = 0 or 1 for predictor variable i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3822f5c1-aec2-44eb-b2fd-7000e8e9756b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_over_hom</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>3838</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65443954</td>\n",
       "      <td>65444209</td>\n",
       "      <td>CZI_abs_1545</td>\n",
       "      <td>255</td>\n",
       "      <td>-</td>\n",
       "      <td>l1</td>\n",
       "      <td>YM2</td>\n",
       "      <td>AF1,YM1,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>80939086</td>\n",
       "      <td>80945263</td>\n",
       "      <td>CZI_abs_1546</td>\n",
       "      <td>6177</td>\n",
       "      <td>-</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF2,AF3,AM1,YF1,YM1,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>82660292</td>\n",
       "      <td>82661886</td>\n",
       "      <td>CZI_abs_1547</td>\n",
       "      <td>1594</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AF2,AM2,YF1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>86275302</td>\n",
       "      <td>86276014</td>\n",
       "      <td>CZI_abs_1548</td>\n",
       "      <td>712</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124088197</td>\n",
       "      <td>124088200</td>\n",
       "      <td>CZI_ins_4331</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF3,AM1,YF1,YF2,YM2</td>\n",
       "      <td>AF2,AM2,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>mixed</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>chr9</td>\n",
       "      <td>126217880</td>\n",
       "      <td>126217897</td>\n",
       "      <td>CZI_ins_4333</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AM1,YM1</td>\n",
       "      <td>AF1,AF2,AF3,YF2,YM2</td>\n",
       "      <td>insertion</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>chr9</td>\n",
       "      <td>127834834</td>\n",
       "      <td>127834845</td>\n",
       "      <td>CZI_ins_4334</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM1,YM2</td>\n",
       "      <td>YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>het</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129426079</td>\n",
       "      <td>129426094</td>\n",
       "      <td>CZI_ins_4335</td>\n",
       "      <td>15</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000</th>\n",
       "      <td>chr9</td>\n",
       "      <td>131003481</td>\n",
       "      <td>131003497</td>\n",
       "      <td>CZI_ins_4336</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>AF3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>het</td>\n",
       "      <td>truncated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5001 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end            id  length strand class  \\\n",
       "0     chr1   56365451   56369289  CZI_abs_1543    3838      +    l1   \n",
       "1     chr1   65443954   65444209  CZI_abs_1545     255      -    l1   \n",
       "2     chr1   80939086   80945263  CZI_abs_1546    6177      -    l1   \n",
       "3     chr1   82660292   82661886  CZI_abs_1547    1594      +    l1   \n",
       "4     chr1   86275302   86276014  CZI_abs_1548     712      +    l1   \n",
       "...    ...        ...        ...           ...     ...    ...   ...   \n",
       "4996  chr9  124088197  124088200  CZI_ins_4331       3      -   alu   \n",
       "4997  chr9  126217880  126217897  CZI_ins_4333      17      +   alu   \n",
       "4998  chr9  127834834  127834845  CZI_ins_4334      11      -   alu   \n",
       "4999  chr9  129426079  129426094  CZI_ins_4335      15      -   alu   \n",
       "5000  chr9  131003481  131003497  CZI_ins_4336      16      +   alu   \n",
       "\n",
       "                                          het  \\\n",
       "0                                         AF3   \n",
       "1                                         YM2   \n",
       "2                                         NaN   \n",
       "3                                         NaN   \n",
       "4                                         NaN   \n",
       "...                                       ...   \n",
       "4996                  AF1,AF3,AM1,YF1,YF2,YM2   \n",
       "4997                                  AM1,YM1   \n",
       "4998      AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM1,YM2   \n",
       "4999                                      NaN   \n",
       "5000  AF1,AF2,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3   \n",
       "\n",
       "                                              hom    me_type  het_over_hom  \\\n",
       "0                             AM1,AM2,AM3,YF2,YM3    absence      0.200000   \n",
       "1                                     AF1,YM1,YM3    absence      0.333333   \n",
       "2                         AF2,AF3,AM1,YF1,YM1,YM3    absence          -inf   \n",
       "3                                 AF1,AF2,AM2,YF1    absence          -inf   \n",
       "4                         AF1,AM1,YF1,YM1,YM2,YM3    absence          -inf   \n",
       "...                                           ...        ...           ...   \n",
       "4996                                  AF2,AM2,YM3  insertion      2.000000   \n",
       "4997                          AF1,AF2,AF3,YF2,YM2  insertion      0.400000   \n",
       "4998                                          YM3  insertion      9.000000   \n",
       "4999  AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3  insertion          -inf   \n",
       "5000                                          AF3  insertion     10.000000   \n",
       "\n",
       "     genotype insertion_category  \n",
       "0         hom          truncated  \n",
       "1         hom          truncated  \n",
       "2         hom        full_length  \n",
       "3         hom          truncated  \n",
       "4         hom          truncated  \n",
       "...       ...                ...  \n",
       "4996    mixed          truncated  \n",
       "4997      hom          truncated  \n",
       "4998      het          truncated  \n",
       "4999      hom          truncated  \n",
       "5000      het          truncated  \n",
       "\n",
       "[5001 rows x 13 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fc59d133-fff3-477c-aea6-a4774b60798b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>me_start</th>\n",
       "      <th>me_end</th>\n",
       "      <th>me_id</th>\n",
       "      <th>me_strand</th>\n",
       "      <th>me_class</th>\n",
       "      <th>het</th>\n",
       "      <th>hom</th>\n",
       "      <th>me_type</th>\n",
       "      <th>het_hom_ratio</th>\n",
       "      <th>genotype</th>\n",
       "      <th>insertion_category</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24066774</td>\n",
       "      <td>24083565</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000230703.1</td>\n",
       "      <td>MYOM3-AS1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>24076520</td>\n",
       "      <td>24076812</td>\n",
       "      <td>CZI_abs_7</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3</td>\n",
       "      <td>AM3,YF2</td>\n",
       "      <td>absence</td>\n",
       "      <td>4.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55915603</td>\n",
       "      <td>55944980</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230250.3</td>\n",
       "      <td>LINC01753</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>55935619</td>\n",
       "      <td>55935930</td>\n",
       "      <td>CZI_abs_19</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56154545</td>\n",
       "      <td>56477687</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000260971.4</td>\n",
       "      <td>ENSG00000260971</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56173433</td>\n",
       "      <td>56524495</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284686.1</td>\n",
       "      <td>ENSG00000284686</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>56365451</td>\n",
       "      <td>56369289</td>\n",
       "      <td>CZI_abs_1543</td>\n",
       "      <td>+</td>\n",
       "      <td>l1</td>\n",
       "      <td>AF3</td>\n",
       "      <td>AM1,AM2,AM3,YF2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>0.2</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>3838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57605847</td>\n",
       "      <td>57606206</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000236888.1</td>\n",
       "      <td>RPS20P5</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>57606184</td>\n",
       "      <td>57606197</td>\n",
       "      <td>CZI_ins_818</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AM3,YF1</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,YM1,YM3</td>\n",
       "      <td>insertion</td>\n",
       "      <td>0.2857142857142857</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>chr8</td>\n",
       "      <td>143636019</td>\n",
       "      <td>143656418</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000183309.12</td>\n",
       "      <td>ZNF623</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143641445</td>\n",
       "      <td>143641764</td>\n",
       "      <td>CZI_abs_1430</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF2,AM3,YM1,YM2</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>chr8</td>\n",
       "      <td>144772224</td>\n",
       "      <td>144787348</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000196378.14</td>\n",
       "      <td>ZNF34</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>144776325</td>\n",
       "      <td>144776913</td>\n",
       "      <td>CZI_abs_1433</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>9.0</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>chr9</td>\n",
       "      <td>35791003</td>\n",
       "      <td>35809732</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159899.16</td>\n",
       "      <td>NPR2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>35803110</td>\n",
       "      <td>35803402</td>\n",
       "      <td>CZI_abs_1464</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>AF1,AM1,AM2,YF1,YF2,YM2,YM3</td>\n",
       "      <td>AF2,AF3,AM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>2.333333333333333</td>\n",
       "      <td>het</td>\n",
       "      <td>full_length</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>chr9</td>\n",
       "      <td>73042745</td>\n",
       "      <td>73056528</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000228024.1</td>\n",
       "      <td>CYP1D1P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>73054613</td>\n",
       "      <td>73054880</td>\n",
       "      <td>CZI_abs_1473</td>\n",
       "      <td>+</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>YM1</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>truncated</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129430652</td>\n",
       "      <td>129451422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000230676.1</td>\n",
       "      <td>ENSG00000230676</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129435084</td>\n",
       "      <td>129435407</td>\n",
       "      <td>CZI_abs_1499</td>\n",
       "      <td>-</td>\n",
       "      <td>alu</td>\n",
       "      <td>.</td>\n",
       "      <td>AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3</td>\n",
       "      <td>absence</td>\n",
       "      <td>-inf</td>\n",
       "      <td>hom</td>\n",
       "      <td>full_length</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr  gene_start   gene_end gene_strand             gene_id  \\\n",
       "0    chr1    24066774   24083565           +   ENSG00000230703.1   \n",
       "1    chr1    55915603   55944980           -   ENSG00000230250.3   \n",
       "2    chr1    56154545   56477687           -   ENSG00000260971.4   \n",
       "3    chr1    56173433   56524495           -   ENSG00000284686.1   \n",
       "4    chr1    57605847   57606206           -   ENSG00000236888.1   \n",
       "..    ...         ...        ...         ...                 ...   \n",
       "122  chr8   143636019  143656418           +  ENSG00000183309.12   \n",
       "123  chr8   144772224  144787348           -  ENSG00000196378.14   \n",
       "124  chr9    35791003   35809732           +  ENSG00000159899.16   \n",
       "125  chr9    73042745   73056528           +   ENSG00000228024.1   \n",
       "126  chr9   129430652  129451422           -   ENSG00000230676.1   \n",
       "\n",
       "           gene_name               gene_type   me_start     me_end  \\\n",
       "0          MYOM3-AS1                  lncRNA   24076520   24076812   \n",
       "1          LINC01753                  lncRNA   55935619   55935930   \n",
       "2    ENSG00000260971                  lncRNA   56365451   56369289   \n",
       "3    ENSG00000284686          protein_coding   56365451   56369289   \n",
       "4            RPS20P5    processed_pseudogene   57606184   57606197   \n",
       "..               ...                     ...        ...        ...   \n",
       "122           ZNF623          protein_coding  143641445  143641764   \n",
       "123            ZNF34          protein_coding  144776325  144776913   \n",
       "124             NPR2          protein_coding   35803110   35803402   \n",
       "125          CYP1D1P  unprocessed_pseudogene   73054613   73054880   \n",
       "126  ENSG00000230676                  lncRNA  129435084  129435407   \n",
       "\n",
       "            me_id me_strand me_class                                  het  \\\n",
       "0       CZI_abs_7         -      alu      AF1,AF2,AF3,AM1,YF1,YM1,YM2,YM3   \n",
       "1      CZI_abs_19         +      alu                                    .   \n",
       "2    CZI_abs_1543         +       l1                                  AF3   \n",
       "3    CZI_abs_1543         +       l1                                  AF3   \n",
       "4     CZI_ins_818         -      alu                              AM3,YF1   \n",
       "..            ...       ...      ...                                  ...   \n",
       "122  CZI_abs_1430         -      alu                                    .   \n",
       "123  CZI_abs_1433         +      alu  AF1,AF2,AF3,AM1,AM2,AM3,YF2,YM2,YM3   \n",
       "124  CZI_abs_1464         -      alu          AF1,AM1,AM2,YF1,YF2,YM2,YM3   \n",
       "125  CZI_abs_1473         +      alu                                    .   \n",
       "126  CZI_abs_1499         -      alu                                    .   \n",
       "\n",
       "                                             hom    me_type  \\\n",
       "0                                        AM3,YF2    absence   \n",
       "1                                            AF1    absence   \n",
       "2                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "3                            AM1,AM2,AM3,YF2,YM3    absence   \n",
       "4                    AF1,AF2,AF3,AM1,AM2,YM1,YM3  insertion   \n",
       "..                                           ...        ...   \n",
       "122                              AF2,AM3,YM1,YM2    absence   \n",
       "123                                          YM1    absence   \n",
       "124                                  AF2,AF3,AM3    absence   \n",
       "125                                          YM1    absence   \n",
       "126  AF1,AF2,AF3,AM1,AM2,AM3,YF1,YF2,YM1,YM2,YM3    absence   \n",
       "\n",
       "          het_hom_ratio genotype insertion_category  length  \n",
       "0                   4.0      het        full_length     292  \n",
       "1                  -inf      hom        full_length     311  \n",
       "2                   0.2      hom          truncated    3838  \n",
       "3                   0.2      hom          truncated    3838  \n",
       "4    0.2857142857142857      hom          truncated      13  \n",
       "..                  ...      ...                ...     ...  \n",
       "122                -inf      hom        full_length     319  \n",
       "123                 9.0      het        full_length     588  \n",
       "124   2.333333333333333      het        full_length     292  \n",
       "125                -inf      hom          truncated     267  \n",
       "126                -inf      hom        full_length     323  \n",
       "\n",
       "[127 rows x 19 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27a2c7fc-3bce-40fb-b959-c1949b0c0a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualise significant effects, by celltype, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2938ba-9163-4cd1-a42c-7c2b2343e46c",
   "metadata": {},
   "source": [
    "# metheQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ef7a6b-16ef-432d-ac80-cd038a611adf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
